{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import csv\n",
    "import os\n",
    "from lxml import etree\n",
    "import xml.etree.ElementTree as ET\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "import time\n",
    "import datetime\n",
    "\n",
    "Entrez.email = \"mnaffairs.intl@gmail.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "aaMapThreeToOne = {'Ala': 'A', 'Arg': 'R', 'Asn': 'N', 'Asp': 'D', 'Cys': 'C',\n",
    "                   'Glu': 'E', 'Gln': 'Q', 'Gly': 'G', 'His': 'H', 'Ile': 'I',\n",
    "                   'Leu': 'L', 'Lys': 'K', 'Met': 'M', 'Phe': 'F', 'Pro': 'P',\n",
    "                   'Ser': 'S', 'Thr': 'T', 'Trp': 'W', 'Tyr': 'Y', 'Val': 'V'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readHumanGenesEC(path):\n",
    "    genes_dict = {}\n",
    "    dup = set()\n",
    "    with open(path, 'r') as f:\n",
    "        f.readline()\n",
    "        reader = csv.reader(f)\n",
    "        for row in reader:\n",
    "            key = row[0]\n",
    "            if key in genes_dict.keys():\n",
    "                dup.add(key)\n",
    "            else:\n",
    "                genes_dict[key] = row[1]\n",
    "    print(f'{len(dup)} genes are duplicated: ', dup)\n",
    "    return genes_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10 genes are duplicated:  {'ERVK-10', '', 'ERVK-18', 'HERVK_113', 'ERVK-7', 'ERVK-19', 'ERVK-6', 'ERVK-8', 'ERVK-25', 'ERVK-9'}\n",
      "4346 genes are imported\n"
     ]
    }
   ],
   "source": [
    "genes_dict = readHumanGenesEC('../data/HumanEnzWithEC.csv')\n",
    "print(f'{len(genes_dict)} genes are imported')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "code_folding": [
     2,
     47,
     107,
     188
    ]
   },
   "outputs": [],
   "source": [
    "class variationHandler(object):\n",
    "    # this class is passed to XMLparser to parse an xml file including info on mutations\n",
    "    def __init__(self, genes_dict):\n",
    "        self.dictlist = {'gene_ID':[], 'gene_name':[], 'clinical_significance': [], 'EC_number': [], 'missense_variation': [], 'NC_accession': [], 'NP_accession': [], 'ClinVar_accession':[], 'Chr': [], 'start':[], 'stop':[], 'referenceAllele':[], 'alternateAllele':[], 'strand': []}\n",
    "        self.genes_dict = genes_dict\n",
    "        self.unnecessary_types = ('inversion', 'copy number gain', 'tandem duplication', 'microsatellite', 'copy number loss', 'distinct chromosomes', 'fusion', 'complex', 'duplication', 'translocation')\n",
    "        self.is_type = False\n",
    "        self.is_type_multi = False  # modified\n",
    "        self.is_simple_allele = False\n",
    "        self.gene_ID = \"\"\n",
    "        self.gene_name = \"\"\n",
    "        self.clinvar_acc = \"\"\n",
    "        self.nc_acc = \"\"\n",
    "        self.np_acc = \"\"\n",
    "        self.change = \"\"\n",
    "        self.chr = \"\"\n",
    "        self.start_num = \"\"\n",
    "        self.stop_num = \"\"\n",
    "        self.referenceAllele = \"\"\n",
    "        self.alternateAllele = \"\"\n",
    "        self.is_GeneList = False\n",
    "        self.ct_gene = 0\n",
    "        self.strand_dict = {}  \n",
    "        self.is_hgvs_grch38 = False\n",
    "        self.ct_np = 0\n",
    "        self.ct_mc = 0  # counter for Molecular Consequence tag\n",
    "        self.is_missense = False\n",
    "        self.is_conflicting = False\n",
    "        self.is_not_provided = False\n",
    "        self.is_no_intpn_add = False\n",
    "        self.is_interpretations = False\n",
    "        self.is_interpretation = False\n",
    "        self.is_description = False\n",
    "        self.is_desc_hist = False\n",
    "        self.intpn = []\n",
    "        self.no_intpn = []\n",
    "        self.intpn_var = []\n",
    "        self.ct = 0\n",
    "        self.ct_missense = 0\n",
    "        self.ct_uncertain = 0\n",
    "        self.ct_conflicting = 0\n",
    "        self.ct_not_provided = 0\n",
    "        self.ct_no_intpn = 0\n",
    "        self.ct_no_intpn_added = 0\n",
    "        self.allele_set = set()\n",
    "        self.allele_set_inside = set()\n",
    "        \n",
    "    def start(self, tag, attrs):\n",
    "        if (tag == 'VariationArchive') and (attrs.get('VariationType').lower() not in self.unnecessary_types):\n",
    "            self.is_type = True\n",
    "            if attrs.get('VariationType').lower() == 'haplotype'  or attrs.get('VariationType').lower() == 'diplotype' or attrs.get('VariationType').lower() == 'compoundheterozygote':\n",
    "                self.is_type_multi = True\n",
    "            self.clinvar_acc = attrs.get('Accession')\n",
    "        if self.is_type:\n",
    "            if tag == 'SimpleAllele':\n",
    "                variation_id = attrs.get('VariationID')\n",
    "                if self.is_type_multi == False:\n",
    "                    if variation_id not in self.allele_set:\n",
    "                        self.allele_set.add(variation_id)\n",
    "                    elif variation_id is not None:\n",
    "                        print('found a duplicated VariationID: ', variation_id)\n",
    "                else:\n",
    "                    if variation_id is not None:\n",
    "                        self.allele_set_inside.add(variation_id)\n",
    "            elif tag == 'GeneList':\n",
    "                self.is_GeneList = True\n",
    "            elif tag == 'Gene' and self.ct_gene == 0:\n",
    "                self.gene_ID = attrs.get('Symbol')\n",
    "                self.gene_name = attrs.get('FullName')\n",
    "                self.ct_gene += 1\n",
    "            elif tag == 'SequenceLocation' and self.is_GeneList == True and self.ct_gene == 1:\n",
    "                if attrs.get('Assembly') == 'GRCh38':\n",
    "                    nc_accession = attrs.get('Accession')\n",
    "                    strand = attrs.get('Strand')\n",
    "                    self.strand_dict[nc_accession] = strand\n",
    "            elif tag == 'SequenceLocation' and self.is_GeneList == False:\n",
    "                if attrs.get('Assembly') == 'GRCh38':\n",
    "                    self.chr = attrs.get('Chr')\n",
    "                    self.start_num = attrs.get('start')\n",
    "                    self.stop_num = attrs.get('stop')\n",
    "                    self.referenceAllele = attrs.get('referenceAlleleVCF')\n",
    "                    self.alternateAllele = attrs.get('alternateAlleleVCF')\n",
    "            elif tag == 'HGVS' and attrs.get('Assembly') == 'GRCh38':\n",
    "                self.is_hgvs_grch38 = True\n",
    "            elif tag == 'NucleotideExpression' and self.is_hgvs_grch38 == True:\n",
    "                self.nc_acc = attrs.get('sequenceAccessionVersion')\n",
    "            elif tag == 'ProteinExpression' and self.ct_np == 0:\n",
    "                self.np_acc = attrs.get('sequenceAccessionVersion')\n",
    "                self.change = attrs.get('change') \n",
    "                if self.np_acc and self.np_acc.startswith('NP'):            \n",
    "                    self.ct_np += 1\n",
    "            elif tag == 'MolecularConsequence' and self.ct_mc == 0:\n",
    "                if attrs.get('Type') and 'missense' in attrs.get('Type').lower():\n",
    "                    self.is_missense = True\n",
    "                    self.ct_missense += 1\n",
    "                self.ct_mc += 1\n",
    "            elif tag == 'Interpretations':\n",
    "                self.is_interpretations = True\n",
    "            elif tag == 'Interpretation':\n",
    "                self.is_interpretation = True\n",
    "            elif tag == 'Description':\n",
    "                self.is_description = True\n",
    "            elif tag == 'DescriptionHistory':\n",
    "                self.is_desc_hist = True\n",
    "            elif tag == 'InterpretedVariation':\n",
    "                self.intpn_var.append(attrs['Accession'])\n",
    "            \n",
    "    def end(self, tag):\n",
    "        if (tag == 'VariationArchive' and self.is_type and (not self.is_type_multi)):\n",
    "            if len(self.intpn) == 1:\n",
    "                clinical_significance = self.intpn[0].lower()\n",
    "                if \"uncertain\" in clinical_significance:\n",
    "                    self.is_uncertain = True\n",
    "                    self.ct_uncertain += 1\n",
    "                elif \"conflicting\" in clinical_significance:\n",
    "                    self.is_conflicting = True\n",
    "                    self.ct_conflicting += 1\n",
    "                elif \"not provided\" in clinical_significance:\n",
    "                    self.is_not_provided = True\n",
    "                    self.ct_not_provided += 1\n",
    "                elif clinical_significance == 'no interpretation for the single variant':\n",
    "                    if set(self.intpn_var).issubset(set(self.dictlist['ClinVar_accession'])):\n",
    "                        self.is_no_intpn_add = True\n",
    "                        self.ct_no_intpn_added += 1\n",
    "                    else:\n",
    "                        self.no_intpn.append(self.clinvar_acc)  #record all clinvar_acc of varations with no interpretaion to check if there are any variants whose haplotype's interpreation cannot be checked because of the order of the xml file\n",
    "                    self.ct_no_intpn += 1\n",
    "            elif len(self.intpn) > 1:\n",
    "                print(f'Interpretaion: {self.intpn}, Accession: {self.clinvar_acc}, count: {self.ct}')\n",
    "            if (self.gene_ID in self.genes_dict.keys()) and self.is_missense and (self.is_uncertain or self.is_conflicting or self.is_not_provided or self.is_no_intpn_add):\n",
    "                try:\n",
    "                    self.change = self.change.split('p.')[1]\n",
    "                    before = aatranlation.get(self.change[0:3])\n",
    "                    after = aatranlation.get(self.change[len(self.change) - 3:len(self.change)])\n",
    "                except: \n",
    "                    before = None\n",
    "                    after = None\n",
    "                if before and after:  # check if both have a value in aa dict\n",
    "                    num = self.change[3:len(self.change) - 3]\n",
    "                    abbreviated_change = before + num + after\n",
    "                    fasta = np.nan\n",
    "                    self.dictlist['gene_ID'].append(self.gene_ID)\n",
    "                    self.dictlist['gene_name'].append(self.gene_name)\n",
    "                    self.dictlist['clinical_significance'].append(clinical_significance)\n",
    "                    self.dictlist['EC_number'].append(self.genes_dict.get(self.gene_ID))\n",
    "                    self.dictlist['missense_variation'].append(abbreviated_change)\n",
    "                    self.dictlist['NP_accession'].append(self.np_acc)\n",
    "                    self.dictlist['NC_accession'].append(self.nc_acc)\n",
    "                    self.dictlist['ClinVar_accession'].append(self.clinvar_acc)\n",
    "                    self.dictlist['Chr'].append(self.chr)\n",
    "                    self.dictlist['start'].append(self.start_num)\n",
    "                    self.dictlist['stop'].append(self.stop_num)\n",
    "                    self.dictlist['referenceAllele'].append(self.referenceAllele)\n",
    "                    self.dictlist['alternateAllele'].append(self.alternateAllele)\n",
    "                    strand = self.strand_dict.get(self.nc_acc)\n",
    "                    self.dictlist['strand'].append(strand)\n",
    "            self.ct_gene = 0             \n",
    "            self.is_missense = False\n",
    "            self.is_uncertain = False\n",
    "            self.is_conflicting = False\n",
    "            self.is_not_provided = False\n",
    "            self.is_no_intpn_add = False\n",
    "            self.ct_np = 0\n",
    "            self.ct_mc = 0\n",
    "            self.intpn = []\n",
    "            self.is_type = False\n",
    "            self.is_type_multi = False\n",
    "            self.ct +=1\n",
    "            if self.ct % 10000 == 0:\n",
    "                print(f'counter: {self.ct}')\n",
    "        if self.is_type:\n",
    "            if tag == 'GeneList':\n",
    "                self.is_GeneList = False\n",
    "            elif tag == 'Interpretations':\n",
    "                self.is_interpretations = False\n",
    "            elif tag == 'Interpretation':\n",
    "                self.is_interpretaion = False\n",
    "            elif tag == 'Description':\n",
    "                self.is_description = False\n",
    "            elif tag == 'DescriptionHistory':\n",
    "                self.is_desc_hist = False\n",
    "            elif tag == 'HGVS' and self.is_hgvs_grch38 == True:\n",
    "                self.is_hgvs_grch38 = False\n",
    "                \n",
    "    def data(self, data):\n",
    "        if (not self.is_simple_allele) and self.is_interpretations and self.is_interpretation and self.is_description and (not self.is_desc_hist):\n",
    "            self.intpn.append(data)\n",
    "            \n",
    "    def close(self):\n",
    "        print(f\"Variations: {self.ct}\")\n",
    "        print(f\"Uncertain Significance: {self.ct_uncertain}\")\n",
    "        print(f\"Conflicting Report: {self.ct_conflicting}\")\n",
    "        print(f\"Not Provided: {self.ct_not_provided}\")\n",
    "        print(f\"No interpretation for the single variant: {self.ct_no_intpn_added}\")\n",
    "        print(f\"Missense: {self.ct_missense}\")\n",
    "        print(f\"No interpretations that aren't included: {self.no_intpn}\")\n",
    "        print(f\"Mutations in the list: {len(self.dictlist['gene_ID'])}\")\n",
    "        print(f\"\")\n",
    "        print('debug: the file is closed')\n",
    "        return self.dictlist, self.allele_set, self.allele_set_inside"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# class variationHandler(object):\n",
    "#     def __init__(self, genes_dict):\n",
    "#         self.dictlist = {'gene_ID':[], 'gene_name':[], 'clinical_significance': [], 'EC_number': [], 'missense_variation': [], 'NC_accession': [], 'NP_accession': [], 'ClinVar_accession':[], 'Chr': [], 'start':[], 'stop':[], 'referenceAllele':[], 'alternateAllele':[], 'strand': []}\n",
    "#         self.genes_dict = genes_dict\n",
    "#         self.unnecessary_types = ('inversion', 'copy number gain', 'tandem duplication', 'microsatellite', 'copy number loss', 'distinct chromosomes', 'fusion', 'complex', 'duplication', 'translocation')\n",
    "#         self.is_type = False\n",
    "#         self.is_simple_allele = False\n",
    "#         self.gene_ID = \"\"\n",
    "#         self.gene_name = \"\"\n",
    "#         self.clinvar_acc = \"\"\n",
    "#         self.nc_acc = \"\"\n",
    "#         self.np_acc = \"\"\n",
    "#         self.change = \"\"\n",
    "#         self.chr = \"\"\n",
    "#         self.start_num = \"\"\n",
    "#         self.stop_num = \"\"\n",
    "#         self.referenceAllele = \"\"\n",
    "#         self.alternateAllele = \"\"\n",
    "#         self.is_GeneList = False\n",
    "#         self.ct_gene = 0\n",
    "#         self.strand_dict = {}  \n",
    "#         self.is_hgvs_grch38 = False\n",
    "#         self.ct_np = 0\n",
    "#         self.ct_mc = 0  # counter for Molecular Consequence tag\n",
    "#         self.is_haplotype = False  # check if a variation is haplotype or not\n",
    "#         self.is_genotype = False  # check if a variation is genotype or not\n",
    "#         self.is_missense = False\n",
    "#         self.is_conflicting = False\n",
    "#         self.is_not_provided = False\n",
    "#         self.is_no_intpn_add = False\n",
    "#         self.is_interpretations = False\n",
    "#         self.is_interpretation = False\n",
    "#         self.is_description = False\n",
    "#         self.is_desc_hist = False\n",
    "#         self.intpn = []\n",
    "#         self.no_intpn = []\n",
    "#         self.intpn_var = []\n",
    "#         self.ct = 0\n",
    "#         self.ct_missense = 0\n",
    "#         self.ct_uncertain = 0\n",
    "#         self.ct_conflicting = 0\n",
    "#         self.ct_not_provided = 0\n",
    "#         self.ct_no_intpn = 0\n",
    "#         self.ct_no_intpn_added = 0\n",
    "#         self.allele_set = set()\n",
    "        \n",
    "#     def start(self, tag, attrs):\n",
    "#         if (tag == 'VariationArchive') and (attrs.get('VariationType').lower() not in self.unnecessary_types):\n",
    "#             self.is_type = True\n",
    "#             if attrs.get('VariationType').lower() == 'haplotype'  or attrs.get('VariationType').lower() == 'diplotype':\n",
    "#                 self.is_haplotype = True\n",
    "#             if attrs.get('VariationType').lower() == 'compoundheterozygote':\n",
    "#                 self.is_genotype = True\n",
    "#             self.clinvar_acc = attrs.get('Accession')\n",
    "#         if self.is_type:\n",
    "#             if tag == 'SimpleAllele':\n",
    "#                 self.is_simple_allele = True\n",
    "#                 variation_id = attrs.get('VariationID')\n",
    "#                 if variation_id not in self.allele_set:\n",
    "#                     self.allele_set.add(variation_id)\n",
    "#                 elif variation_id is not None:\n",
    "#                     print('found a duplicated VariationID: ', variation_id)\n",
    "#             elif tag == 'GeneList':\n",
    "#                 self.is_GeneList = True\n",
    "#             elif tag == 'Gene' and self.ct_gene == 0:\n",
    "#                 self.gene_ID = attrs.get('Symbol')\n",
    "#                 self.gene_name = attrs.get('FullName')\n",
    "#                 self.ct_gene += 1\n",
    "#             elif tag == 'SequenceLocation' and self.is_GeneList == True and self.ct_gene == 1:\n",
    "#                 if attrs.get('Assembly') == 'GRCh38':\n",
    "#                     nc_accession = attrs.get('Accession')\n",
    "#                     strand = attrs.get('Strand')\n",
    "#                     self.strand_dict[nc_accession] = strand\n",
    "#             elif tag == 'SequenceLocation' and self.is_GeneList == False:\n",
    "#                 if attrs.get('Assembly') == 'GRCh38':\n",
    "#                     self.chr = attrs.get('Chr')\n",
    "#                     self.start_num = attrs.get('start')\n",
    "#                     self.stop_num = attrs.get('stop')\n",
    "#                     self.referenceAllele = attrs.get('referenceAlleleVCF')\n",
    "#                     self.alternateAllele = attrs.get('alternateAlleleVCF')\n",
    "#             elif tag == 'HGVS' and attrs.get('Assembly') == 'GRCh38':\n",
    "#                 self.is_hgvs_grch38 = True\n",
    "#             elif tag == 'NucleotideExpression' and self.is_hgvs_grch38 == True:\n",
    "#                 self.nc_acc = attrs.get('sequenceAccessionVersion')\n",
    "#             elif tag == 'ProteinExpression' and self.ct_np == 0:\n",
    "#                 self.np_acc = attrs.get('sequenceAccessionVersion')\n",
    "#                 self.change = attrs.get('change') \n",
    "#                 if self.np_acc and self.np_acc.startswith('NP'):            \n",
    "#                     self.ct_np += 1\n",
    "#             elif tag == 'MolecularConsequence' and self.ct_mc == 0:\n",
    "#                 if attrs.get('Type') and 'missense' in attrs.get('Type').lower():\n",
    "#                     self.is_missense = True\n",
    "#                     self.ct_missense += 1\n",
    "#                 self.ct_mc += 1\n",
    "#             elif tag == 'Interpretations':\n",
    "#                 self.is_interpretations = True\n",
    "#             elif tag == 'Interpretation':\n",
    "#                 self.is_interpretation = True\n",
    "#             elif tag == 'Description':\n",
    "#                 self.is_description = True\n",
    "#             elif tag == 'DescriptionHistory':\n",
    "#                 self.is_desc_hist = True\n",
    "#             elif tag == 'InterpretedVariation':\n",
    "#                 self.intpn_var.append(attrs['Accession'])\n",
    "            \n",
    "#     def end(self, tag):\n",
    "#         if (tag == 'VariationArchive' and self.is_type) or ((self.is_haplotype or self.is_genotype) and tag == 'SimpleAllele'):\n",
    "#             if len(self.intpn) == 1:\n",
    "#                 clinical_significance = self.intpn[0].lower()\n",
    "#                 if \"uncertain\" in clinical_significance:\n",
    "#                     self.is_uncertain = True\n",
    "#                     self.ct_uncertain += 1\n",
    "#                 elif \"conflicting\" in clinical_significance:\n",
    "#                     self.is_conflicting = True\n",
    "#                     self.ct_conflicting += 1\n",
    "#                 elif \"not provided\" in clinical_significance:\n",
    "#                     self.is_not_provided = True\n",
    "#                     self.ct_not_provided += 1\n",
    "#                 elif clinical_significance == 'no interpretation for the single variant':\n",
    "#                     if set(self.intpn_var).issubset(set(self.dictlist['ClinVar_accession'])):\n",
    "#                         self.is_no_intpn_add = True\n",
    "#                         self.ct_no_intpn_added += 1\n",
    "#                     else:\n",
    "#                         self.no_intpn.append(self.clinvar_acc)  #record all clinvar_acc of varations with no interpretaion to check if there are any variants whose haplotype's interpreation cannot be checked because of the order of the xml file\n",
    "#                     self.ct_no_intpn += 1\n",
    "#             elif len(self.intpn) > 1:\n",
    "#                 print(f'Interpretaion: {self.intpn}, Accession: {self.clinvar_acc}, count: {self.ct}')\n",
    "#             if (self.gene_ID in self.genes_dict.keys()) and self.is_missense and (self.is_uncertain or self.is_conflicting or self.is_not_provided or self.is_no_intpn_add):\n",
    "#                 try:\n",
    "#                     self.change = self.change.split('p.')[1]\n",
    "#                     before = aatranlation.get(self.change[0:3])\n",
    "#                     after = aatranlation.get(self.change[len(self.change) - 3:len(self.change)])\n",
    "#                 except: \n",
    "#                     before = None\n",
    "#                     after = None\n",
    "#                 if before and after:  # check if both have a value in aa dict\n",
    "#                     num = self.change[3:len(self.change) - 3]\n",
    "#                     abbreviated_change = before + num + after\n",
    "#                     fasta = np.nan\n",
    "#                     self.dictlist['gene_ID'].append(self.gene_ID)\n",
    "#                     self.dictlist['gene_name'].append(self.gene_name)\n",
    "#                     self.dictlist['clinical_significance'].append(clinical_significance)\n",
    "#                     self.dictlist['EC_number'].append(self.genes_dict.get(self.gene_ID))\n",
    "#                     self.dictlist['missense_variation'].append(abbreviated_change)\n",
    "#                     self.dictlist['NP_accession'].append(self.np_acc)\n",
    "#                     self.dictlist['NC_accession'].append(self.nc_acc)\n",
    "#                     self.dictlist['ClinVar_accession'].append(self.clinvar_acc)\n",
    "#                     self.dictlist['Chr'].append(self.chr)\n",
    "#                     self.dictlist['start'].append(self.start_num)\n",
    "#                     self.dictlist['stop'].append(self.stop_num)\n",
    "#                     self.dictlist['referenceAllele'].append(self.referenceAllele)\n",
    "#                     self.dictlist['alternateAllele'].append(self.alternateAllele)\n",
    "#                     strand = self.strand_dict.get(self.nc_acc)\n",
    "#                     self.dictlist['strand'].append(strand)\n",
    "#             self.ct_gene = 0             \n",
    "#             self.is_missense = False\n",
    "#             self.is_uncertain = False\n",
    "#             self.is_conflicting = False\n",
    "#             self.is_not_provided = False\n",
    "#             self.is_no_intpn_add = False\n",
    "#             self.ct_np = 0\n",
    "#             self.ct_mc = 0\n",
    "#             self.intpn = []\n",
    "#         if self.is_type:\n",
    "#             if tag == 'SimpleAllele':\n",
    "#                 self.is_simple_allele = False\n",
    "#             elif tag == 'GeneList':\n",
    "#                 self.is_GeneList = False\n",
    "#             elif tag == 'Interpretations':\n",
    "#                 self.is_interpretations = False\n",
    "#             elif tag == 'Interpretation':\n",
    "#                 self.is_interpretaion = False\n",
    "#             elif tag == 'Description':\n",
    "#                 self.is_description = False\n",
    "#             elif tag == 'DescriptionHistory':\n",
    "#                 self.is_desc_hist = False\n",
    "#             elif tag == 'HGVS' and self.is_hgvs_grch38 == True:\n",
    "#                 self.is_hgvs_grch38 = False\n",
    "#         if tag == 'VariationArchive':\n",
    "#             self.is_type = False\n",
    "#             self.is_haplotype = False\n",
    "#             self.is_genotype = False\n",
    "#             self.ct +=1\n",
    "#             if self.ct % 10000 == 0:\n",
    "#                 print(f'counter: {self.ct}')\n",
    "                \n",
    "#     def data(self, data):\n",
    "#         if (not self.is_simple_allele) and self.is_interpretations and self.is_interpretation and self.is_description and (not self.is_desc_hist):\n",
    "#             self.intpn.append(data)\n",
    "            \n",
    "#     def close(self):\n",
    "#         print(f\"Variations: {self.ct}\")\n",
    "#         print(f\"Uncertain Significance: {self.ct_uncertain}\")\n",
    "#         print(f\"Conflicting Report: {self.ct_conflicting}\")\n",
    "#         print(f\"Not Provided: {self.ct_not_provided}\")\n",
    "#         print(f\"No interpretation for the single variant: {self.ct_no_intpn_added}\")\n",
    "#         print(f\"Missense: {self.ct_missense}\")\n",
    "#         print(f\"No interpretations that aren't included: {self.no_intpn}\")\n",
    "#         print(f\"Mutations in the list: {len(self.dictlist['gene_ID'])}\")\n",
    "#         print('debug: the file is closed')\n",
    "#         return self.dictlist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# # read xml file of variations from ClinVar\n",
    "# # return dataframe and write to a csv file\n",
    "# def readClinVarVariationsXML(input_path, output_path, gene_set):\n",
    "#     print('debug: start parcing')\n",
    "#     parser = etree.XMLParser(target=variationHandler(gene_set))\n",
    "#     data = etree.parse(input_path, parser)\n",
    "#     df = pd.DataFrame(data)\n",
    "#     df.to_csv(output_path, index = False, header = True)\n",
    "#     return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# read xml file of variations from ClinVar\n",
    "# return dataframe and write to a csv file\n",
    "def readClinVarVariationsXML(input_path, output_path, gene_set):\n",
    "    print('debug: start parcing')\n",
    "    parser = etree.XMLParser(target=variationHandler(gene_set))\n",
    "    data, allele_set, allele_set_inside = etree.parse(input_path, parser)\n",
    "    df = pd.DataFrame(data)\n",
    "    df.to_csv(output_path, index = False, header = True)\n",
    "    return df, allele_set, allele_set_inside"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "debug: start parcing\n",
      "Variations: 355\n",
      "Uncertain Significance: 190\n",
      "Conflicting Report: 0\n",
      "Not Provided: 0\n",
      "No interpretation for the single variant: 0\n",
      "Missense: 204\n",
      "No interpretations that aren't included: []\n",
      "Mutations in the list: 16\n",
      "\n",
      "debug: the file is closed\n",
      "356 125\n",
      "{'242771', '830294', '830244', '830272', '830248', '830301', '830315', '830350', '830259', '830252', '830303', '830335', '830260', '830306', '830300', '830280', '830239', '830342', '830311', '830363', '830240', '830341', '830351', '830245', '830289', '830284', '830302', '830353', '830362', '830288', '830331', '830276', '830254', '830281', '830359', '830354', '830269', '830282', '830366', '830317', '39379', '830278', '830242', '830286', '830295', '830367', '830290', '830298', '830285', '830268', '830352', '830304', '830287', '830348', '830296', '830291', '830270', '830305', '830333', '830297', '830261', '830277', '830299', '830236', '830283', '830334', '830337', '830271', '830339', '830365', '830275', '830265', '830371', '830316', '830263', '830264', '830237', '830249', '830349', '830357', '830338', '830251', '830361', '830246', '830330', '830368', '830255', '830356', '830257', '830372', '830279', '830343', '830258', '830308', '830247', '830243', '830267', '830318', '830364', '830253', '830360', '214016', '830312', '830347', '830292', '830344', '830262', '830332', '830358', '830340', '830309', '830238', '830256', '830355', '830266', '830345', '830250', '242701', '830274', '830293', '830241', '830273', '830336', '830307', '830346'}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>EC_number</th>\n",
       "      <th>missense_variation</th>\n",
       "      <th>NC_accession</th>\n",
       "      <th>NP_accession</th>\n",
       "      <th>ClinVar_accession</th>\n",
       "      <th>Chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>referenceAllele</th>\n",
       "      <th>alternateAllele</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHD2</td>\n",
       "      <td>chromodomain helicase DNA binding protein 2</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>3.6.4.12</td>\n",
       "      <td>R178G</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>NP_001036037.1</td>\n",
       "      <td>VCV000829865</td>\n",
       "      <td>15</td>\n",
       "      <td>92937606</td>\n",
       "      <td>92937606</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SPATA5</td>\n",
       "      <td>spermatogenesis associated 5</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>3.6.4.10</td>\n",
       "      <td>P813S</td>\n",
       "      <td>NC_000004.12</td>\n",
       "      <td>NP_001332785.1</td>\n",
       "      <td>VCV000829872</td>\n",
       "      <td>4</td>\n",
       "      <td>123256115</td>\n",
       "      <td>123256115</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CFB</td>\n",
       "      <td>complement factor B</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>3.4.21.47</td>\n",
       "      <td>R143C</td>\n",
       "      <td>NC_000006.12</td>\n",
       "      <td>NP_001701.2</td>\n",
       "      <td>VCV000829882</td>\n",
       "      <td>6</td>\n",
       "      <td>31947135</td>\n",
       "      <td>31947135</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EYA1</td>\n",
       "      <td>EYA transcriptional coactivator and phosphatase 1</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>3.1.3.16; 3.1.3.48</td>\n",
       "      <td>D330Y</td>\n",
       "      <td>NC_000008.11</td>\n",
       "      <td>NP_000494.2</td>\n",
       "      <td>VCV000829887</td>\n",
       "      <td>8</td>\n",
       "      <td>71269802</td>\n",
       "      <td>71269802</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CFI</td>\n",
       "      <td>complement factor I</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>3.4.21.45</td>\n",
       "      <td>C284G</td>\n",
       "      <td>NC_000004.12</td>\n",
       "      <td>NP_000195.3</td>\n",
       "      <td>VCV000829894</td>\n",
       "      <td>4</td>\n",
       "      <td>109760303</td>\n",
       "      <td>109760303</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_ID                                          gene_name  \\\n",
       "0    CHD2        chromodomain helicase DNA binding protein 2   \n",
       "1  SPATA5                       spermatogenesis associated 5   \n",
       "2     CFB                                complement factor B   \n",
       "3    EYA1  EYA transcriptional coactivator and phosphatase 1   \n",
       "4     CFI                                complement factor I   \n",
       "\n",
       "    clinical_significance           EC_number missense_variation  \\\n",
       "0  uncertain significance            3.6.4.12              R178G   \n",
       "1  uncertain significance            3.6.4.10              P813S   \n",
       "2  uncertain significance           3.4.21.47              R143C   \n",
       "3  uncertain significance  3.1.3.16; 3.1.3.48              D330Y   \n",
       "4  uncertain significance           3.4.21.45              C284G   \n",
       "\n",
       "   NC_accession    NP_accession ClinVar_accession Chr      start       stop  \\\n",
       "0  NC_000015.10  NP_001036037.1      VCV000829865  15   92937606   92937606   \n",
       "1  NC_000004.12  NP_001332785.1      VCV000829872   4  123256115  123256115   \n",
       "2  NC_000006.12     NP_001701.2      VCV000829882   6   31947135   31947135   \n",
       "3  NC_000008.11     NP_000494.2      VCV000829887   8   71269802   71269802   \n",
       "4  NC_000004.12     NP_000195.3      VCV000829894   4  109760303  109760303   \n",
       "\n",
       "  referenceAllele alternateAllele strand  \n",
       "0               A               G      +  \n",
       "1               C               T      +  \n",
       "2               C               T      +  \n",
       "3               C               A      -  \n",
       "4               A               C      -  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## parse a part of the xml file\n",
    "xmlfile = '../data/clinvartail_5.xml'\n",
    "out_path = '../data/MM_enzyme_short.csv'\n",
    "df_short, allele_set, allele_set_inside = readClinVarVariationsXML(xmlfile, out_path, genes_dict)\n",
    "print(len(allele_set), len(allele_set_inside))\n",
    "print(allele_set_inside - allele_set)\n",
    "df_short.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "code_folding": [],
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "debug: start parcing\n",
      "Variations: 2458\n",
      "Uncertain Significance: 65\n",
      "Conflicting Report: 3\n",
      "Not Provided: 17\n",
      "No interpretation for the single variant: 0\n",
      "Missense: 407\n",
      "No interpretations that aren't included: []\n",
      "Mutations in the list: 12\n",
      "\n",
      "debug: the file is closed\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-103-587a6d083773>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mxmlfile\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'../data/ClinVarVariationRelease_00-latest_weekly.xml'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mout_path\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'../data/MM_enzyme.csv'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mdf_0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mallele_set\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mallele_set_inside\u001b[0m  \u001b[0;34m=\u001b[0m \u001b[0mreadClinVarVariationsXML\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mxmlfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgenes_dict\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mallele_set\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mallele_set_inside\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mallele_set_inside\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mallele_set\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-98-1137bcd055ed>\u001b[0m in \u001b[0;36mreadClinVarVariationsXML\u001b[0;34m(input_path, output_path, gene_set)\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'debug: start parcing'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0metree\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mXMLParser\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtarget\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvariationHandler\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgene_set\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m     \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mallele_set\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mallele_set_inside\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0metree\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparser\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m     \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m     \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutput_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32msrc/lxml/etree.pyx\u001b[0m in \u001b[0;36mlxml.etree.parse\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32msrc/lxml/parser.pxi\u001b[0m in \u001b[0;36mlxml.etree._parseDocument\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32msrc/lxml/parser.pxi\u001b[0m in \u001b[0;36mlxml.etree._parseDocumentFromURL\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32msrc/lxml/parser.pxi\u001b[0m in \u001b[0;36mlxml.etree._parseDocFromFile\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32msrc/lxml/parser.pxi\u001b[0m in \u001b[0;36mlxml.etree._BaseParser._parseDocFromFile\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32msrc/lxml/parsertarget.pxi\u001b[0m in \u001b[0;36mlxml.etree._TargetParserContext._handleParseResultDoc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32msrc/lxml/parsertarget.pxi\u001b[0m in \u001b[0;36mlxml.etree._TargetParserContext._handleParseResultDoc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32msrc/lxml/etree.pyx\u001b[0m in \u001b[0;36mlxml.etree._ExceptionContext._raise_if_stored\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32msrc/lxml/saxparser.pxi\u001b[0m in \u001b[0;36mlxml.etree._handleSaxEnd\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32msrc/lxml/parsertarget.pxi\u001b[0m in \u001b[0;36mlxml.etree._PythonSaxParserTarget._handleSaxEnd\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m<ipython-input-97-1d290a7d62fb>\u001b[0m in \u001b[0;36mend\u001b[0;34m(self, tag)\u001b[0m\n\u001b[1;32m    177\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0mtag\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'Description'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    178\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_description\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 179\u001b[0;31m             \u001b[0;32melif\u001b[0m \u001b[0mtag\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'DescriptionHistory'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    180\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_desc_hist\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    181\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0mtag\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'HGVS'\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_hgvs_grch38\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "## parse the entire xml file\n",
    "xmlfile = '../data/ClinVarVariationRelease_00-latest_weekly.xml'\n",
    "out_path = '../data/MM_enzyme.csv'\n",
    "df_0, allele_set, allele_set_inside  = readClinVarVariationsXML(xmlfile, out_path, genes_dict)\n",
    "print(len(allele_set), len(allele_set_inside))\n",
    "print(allele_set_inside - allele_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "568\n"
     ]
    }
   ],
   "source": [
    "ls = ['VCV000012173', 'VCV000025511', 'VCV000132775', 'VCV000132776', 'VCV000242384', 'VCV000242385', 'VCV000242419', 'VCV000242420', 'VCV000242421', 'VCV000242427', 'VCV000242431', 'VCV000242432', 'VCV000242433', 'VCV000242435', 'VCV000242436', 'VCV000242439', 'VCV000242440', 'VCV000242474', 'VCV000242475', 'VCV000242476', 'VCV000242486', 'VCV000242487', 'VCV000242492', 'VCV000242504', 'VCV000242512', 'VCV000242514', 'VCV000242522', 'VCV000242525', 'VCV000242526', 'VCV000242533', 'VCV000242534', 'VCV000242571', 'VCV000242574', 'VCV000242575', 'VCV000242590', 'VCV000242591', 'VCV000242618', 'VCV000242619', 'VCV000242636', 'VCV000242637', 'VCV000242638', 'VCV000242639', 'VCV000242652', 'VCV000242653', 'VCV000242654', 'VCV000242667', 'VCV000242668', 'VCV000242669', 'VCV000242670', 'VCV000242673', 'VCV000242674', 'VCV000242688', 'VCV000242705', 'VCV000242709', 'VCV000242714', 'VCV000242717', 'VCV000242718', 'VCV000242720', 'VCV000242723', 'VCV000242726', 'VCV000242730', 'VCV000242732', 'VCV000242738', 'VCV000242760', 'VCV000242761', 'VCV000242776', 'VCV000242782', 'VCV000242785', 'VCV000242789', 'VCV000242792', 'VCV000242794', 'VCV000242798', 'VCV000242799', 'VCV000242801', 'VCV000242809', 'VCV000242813', 'VCV000242820', 'VCV000264658', 'VCV000264659', 'VCV000264660', 'VCV000264668', 'VCV000264669', 'VCV000264672', 'VCV000264675', 'VCV000267283', 'VCV000267284', 'VCV000267287', 'VCV000267288', 'VCV000267289', 'VCV000267742', 'VCV000267743', 'VCV000279614', 'VCV000279615', 'VCV000374405', 'VCV000374406', 'VCV000375324', 'VCV000375325', 'VCV000375598', 'VCV000375599', 'VCV000375758', 'VCV000375759', 'VCV000393290', 'VCV000393291', 'VCV000393293', 'VCV000393294', 'VCV000393295', 'VCV000393296', 'VCV000417674', 'VCV000417675', 'VCV000424698', 'VCV000428591', 'VCV000428592', 'VCV000429193', 'VCV000429194', 'VCV000430661', 'VCV000431059', 'VCV000431060', 'VCV000440777', 'VCV000441235', 'VCV000446382', 'VCV000446480', 'VCV000478852', 'VCV000478887', 'VCV000478888', 'VCV000478913', 'VCV000478918', 'VCV000478919', 'VCV000487345', 'VCV000487347', 'VCV000545635', 'VCV000549834', 'VCV000560198', 'VCV000560199', 'VCV000560201', 'VCV000617633', 'VCV000617662', 'VCV000617663', 'VCV000624596', 'VCV000624597', 'VCV000624598', 'VCV000624607', 'VCV000624608', 'VCV000624613', 'VCV000635091', 'VCV000635093', 'VCV000635827', 'VCV000635828', 'VCV000689496', 'VCV000242681', 'VCV000624611', 'VCV000624612', 'VCV000427808', 'VCV000427809', 'VCV000478832', 'VCV000478838', 'VCV000478850', 'VCV000478851', 'VCV000242503', 'VCV000370036', 'VCV000242498', 'VCV000242516', 'VCV000242620', 'VCV000242700', 'VCV000242734', 'VCV000242770', 'VCV000242816', 'VCV000242751', 'VCV000446195', 'VCV000496587', 'VCV000242562', 'VCV000242563', 'VCV000242564', 'VCV000242593', 'VCV000242594', 'VCV000242596', 'VCV000242597', 'VCV000635094', 'VCV000635095', 'VCV000635096', 'VCV000635097', 'VCV000560533', 'VCV000585242', 'VCV000242472', 'VCV000242704', 'VCV000242708', 'VCV000242737', 'VCV000242746', 'VCV000242754', 'VCV000242775', 'VCV000242777', 'VCV000242807', 'VCV000242819', 'VCV000242586', 'VCV000242588', 'VCV000242750', 'VCV000242797', 'VCV000446183', 'VCV000446188', 'VCV000692326', 'VCV000692327', 'VCV000242396', 'VCV000242500', 'VCV000242501', 'VCV000242582', 'VCV000242583', 'VCV000242394', 'VCV000242395', 'VCV000242404', 'VCV000242406', 'VCV000242407', 'VCV000242426', 'VCV000242428', 'VCV000242598', 'VCV000242599', 'VCV000242624', 'VCV000242690', 'VCV000242489', 'VCV000242755', 'VCV000242756', 'VCV000242821', 'VCV000267280', 'VCV000267452', 'VCV000267453', 'VCV000267454', 'VCV000267455', 'VCV000267456', 'VCV000267457', 'VCV000267458', 'VCV000267459', 'VCV000267460', 'VCV000267461', 'VCV000267462', 'VCV000267463', 'VCV000375319', 'VCV000375320', 'VCV000478828', 'VCV000478845', 'VCV000478899', 'VCV000518442', 'VCV000561162', 'VCV000624606', 'VCV000375326', 'VCV000375327', 'VCV000631475', 'VCV000631476', 'VCV000039358', 'VCV000242399', 'VCV000242401', 'VCV000242402', 'VCV000242403', 'VCV000242446', 'VCV000242447', 'VCV000242496', 'VCV000242497', 'VCV000242505', 'VCV000242506', 'VCV000242510', 'VCV000242530', 'VCV000242531', 'VCV000242572', 'VCV000242573', 'VCV000242576', 'VCV000242579', 'VCV000242610', 'VCV000242611', 'VCV000242616', 'VCV000242617', 'VCV000242622', 'VCV000242623', 'VCV000242655', 'VCV000242657', 'VCV000242658', 'VCV000242659', 'VCV000242660', 'VCV000242661', 'VCV000242703', 'VCV000242741', 'VCV000242742', 'VCV000242743', 'VCV000242774', 'VCV000267744', 'VCV000267745', 'VCV000375728', 'VCV000375730', 'VCV000417816', 'VCV000438270', 'VCV000438271', 'VCV000438272', 'VCV000487089', 'VCV000488063', 'VCV000488425', 'VCV000488426', 'VCV000488427', 'VCV000627560', 'VCV000627561', 'VCV000635090', 'VCV000635092', 'VCV000684769', 'VCV000684770', 'VCV000242666', 'VCV000560532', 'VCV000021159', 'VCV000050891', 'VCV000050892', 'VCV000050893', 'VCV000050894', 'VCV000050895', 'VCV000050896', 'VCV000242390', 'VCV000242408', 'VCV000242438', 'VCV000242459', 'VCV000242462', 'VCV000242465', 'VCV000242467', 'VCV000242468', 'VCV000242494', 'VCV000242540', 'VCV000242544', 'VCV000242546', 'VCV000242547', 'VCV000242549', 'VCV000242551', 'VCV000242552', 'VCV000242553', 'VCV000242554', 'VCV000242555', 'VCV000242556', 'VCV000242560', 'VCV000242612', 'VCV000242613', 'VCV000242650', 'VCV000242651', 'VCV000242719', 'VCV000242724', 'VCV000242729', 'VCV000242795', 'VCV000242382', 'VCV000242443', 'VCV000242444', 'VCV000242445', 'VCV000242509', 'VCV000242565', 'VCV000242566', 'VCV000242567', 'VCV000242568', 'VCV000242569', 'VCV000242570', 'VCV000242684', 'VCV000242686', 'VCV000242687', 'VCV000242749', 'VCV000242796', 'VCV000264664', 'VCV000264674', 'VCV000267281', 'VCV000267282', 'VCV000402119', 'VCV000430836', 'VCV000430847', 'VCV000430848', 'VCV000429025', 'VCV000429026', 'VCV000430872', 'VCV000430873', 'VCV000433543', 'VCV000433545', 'VCV000433546', 'VCV000433556', 'VCV000433557', 'VCV000446487', 'VCV000444001', 'VCV000444002', 'VCV000446191', 'VCV000545510', 'VCV000545623', 'VCV000545624', 'VCV000559423', 'VCV000559434', 'VCV000624609', 'VCV000624610', 'VCV000624615', 'VCV000636258', 'VCV000636259', 'VCV000637951', 'VCV000691527', 'VCV000691528', 'VCV000694741', 'VCV000694742', 'VCV000242645', 'VCV000242646', 'VCV000242513', 'VCV000242627', 'VCV000242628', 'VCV000431454', 'VCV000560531', 'VCV000585008', 'VCV000691963', 'VCV000242470', 'VCV000242471', 'VCV000242478', 'VCV000242604', 'VCV000242728', 'VCV000242769', 'VCV000242803', 'VCV000264663', 'VCV000264666', 'VCV000264667', 'VCV000268096', 'VCV000268097', 'VCV000393289', 'VCV000393292', 'VCV000397632', 'VCV000478870', 'VCV000478872', 'VCV000478884', 'VCV000478890', 'VCV000478904', 'VCV000478905', 'VCV000487346', 'VCV000559420', 'VCV000559421', 'VCV000590257', 'VCV000590258', 'VCV000624599', 'VCV000624600', 'VCV000624601', 'VCV000624602', 'VCV000624603', 'VCV000624604', 'VCV000624605', 'VCV000684385', 'VCV000684386', 'VCV000813875', 'VCV000813876', 'VCV000242469', 'VCV000242519', 'VCV000007169', 'VCV000208553', 'VCV000208554', 'VCV000242450', 'VCV000242485', 'VCV000242495', 'VCV000242671', 'VCV000242791', 'VCV000264670', 'VCV000264671', 'VCV000267740', 'VCV000267741', 'VCV000373892', 'VCV000375250', 'VCV000375732', 'VCV000430722', 'VCV000437930', 'VCV000440459', 'VCV000446189', 'VCV000478878', 'VCV000559943', 'VCV000559944', 'VCV000818230', 'VCV000818231', 'VCV000242721', 'VCV000478834', 'VCV000487077', 'VCV000487078', 'VCV000487079', 'VCV000487080', 'VCV000487081', 'VCV000827582', 'VCV000827583', 'VCV000242762', 'VCV000242693', 'VCV000242641', 'VCV000242521', 'VCV000242381', 'VCV000242380', 'VCV000375321', 'VCV000375323', 'VCV000446182', 'VCV000634859', 'VCV000634861', 'VCV000634864', 'VCV000634865', 'VCV000634875', 'VCV000634876', 'VCV000634877', 'VCV000634879', 'VCV000634882', 'VCV000830334', 'VCV000830335', 'VCV000830336', 'VCV000830337', 'VCV000830338', 'VCV000830339', 'VCV000830340', 'VCV000830341', 'VCV000830342', 'VCV000830343', 'VCV000830344', 'VCV000830345', 'VCV000830346', 'VCV000830347', 'VCV000830348', 'VCV000830349', 'VCV000830350', 'VCV000830351', 'VCV000830352', 'VCV000830353', 'VCV000830354', 'VCV000024962', 'VCV000024987', 'VCV000050883', 'VCV000050885', 'VCV000132156', 'VCV000242379', 'VCV000242388', 'VCV000242389', 'VCV000242412', 'VCV000242413', 'VCV000242416', 'VCV000242452', 'VCV000242460', 'VCV000242480', 'VCV000242481', 'VCV000242491', 'VCV000242493', 'VCV000242511', 'VCV000242523', 'VCV000242528', 'VCV000242581', 'VCV000242609', 'VCV000242614', 'VCV000242676', 'VCV000242677', 'VCV000242678', 'VCV000242715', 'VCV000242752', 'VCV000242788', 'VCV000242802', 'VCV000242814', 'VCV000264661', 'VCV000267276', 'VCV000375322', 'VCV000375729', 'VCV000375757', 'VCV000424703', 'VCV000425600', 'VCV000425651', 'VCV000440737', 'VCV000440739', 'VCV000446185', 'VCV000446197', 'VCV000478854', 'VCV000487488', 'VCV000487489', 'VCV000487490', 'VCV000487491', 'VCV000493610', 'VCV000549832', 'VCV000549833', 'VCV000549835', 'VCV000549836', 'VCV000549837', 'VCV000549838', 'VCV000618966', 'VCV000626312', 'VCV000691964', 'VCV000830372']\n",
    "print(len(ls))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "debug: start parcing\n",
      "Interpretaion: ['drug response3', 'drug response4'], Accession: VCV000633857, count: 0\n",
      "Variations: 1\n",
      "Uncertain Significance: 0\n",
      "Conflicting Report: 0\n",
      "Not Provided: 0\n",
      "No interpretation for the single variant: 0\n",
      "Missense: 0\n",
      "No interpretations that aren't included: []\n",
      "Mutations in the list: 0\n",
      "debug: the file is closed\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>EC_number</th>\n",
       "      <th>missense_variation</th>\n",
       "      <th>NC_accession</th>\n",
       "      <th>NP_accession</th>\n",
       "      <th>ClinVar_accession</th>\n",
       "      <th>Chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>referenceAllele</th>\n",
       "      <th>alternateAllele</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gene_ID, gene_name, clinical_significance, EC_number, missense_variation, NC_accession, NP_accession, ClinVar_accession, Chr, start, stop, referenceAllele, alternateAllele, strand]\n",
       "Index: []"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # this column is a test for a certain variation I want to look over\n",
    "# subnode = '../data/subnode.xml'\n",
    "# temp_path = '../data/temp.csv'\n",
    "# df_temp = readClinVarVariationsXML(subnode, temp_path, genes_dict)\n",
    "# df_temp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "code_folding": [
     2,
     8,
     26,
     39
    ]
   },
   "outputs": [],
   "source": [
    "# creates an subnode xml file when you want to look over a certain mutation\n",
    "class variationHandlerSpecific(object):\n",
    "    def __init__(self, accession):\n",
    "        self.is_accession = False\n",
    "        self.accession = accession\n",
    "        self.ct = 0\n",
    "        print(self.accession)\n",
    "        \n",
    "    def start(self, tag, attrs):\n",
    "        global WFILE\n",
    "        if (tag == 'VariationArchive') and (attrs.get('Accession') == self.accession):\n",
    "            self.is_accession = True\n",
    "            print('The specific variation is found: ' + str(self.ct))\n",
    "        if self.is_accession:\n",
    "            if len(attrs.keys()) == 0:\n",
    "                WFILE.write('<' + tag)\n",
    "            else:   \n",
    "                for i, t in enumerate(attrs.keys()):\n",
    "                    if i == 0:\n",
    "                        WFILE.write('<' + tag + ' ')\n",
    "                    elif i != len(attrs.keys()) - 1:\n",
    "                        WFILE.write(t + '=\"' + attrs.get(t) + '\"' + ' ')\n",
    "                    else:\n",
    "                        WFILE.write(t + '=\"' + attrs.get(t) + '\"')\n",
    "            WFILE.write('>')\n",
    "            \n",
    "    def end(self, tag):\n",
    "        global WFILE\n",
    "        if self.is_accession:\n",
    "            WFILE.write('</' + tag + '>')\n",
    "        if tag == 'VariationArchive':\n",
    "            if self.ct % 10000 == 0:\n",
    "                print(self.ct)\n",
    "            self.ct += 1\n",
    "        if self.is_accession and tag == 'VariationArchive':\n",
    "            self.is_accession = False\n",
    "            WFILE.close()\n",
    "            print('The subnode file is completed')\n",
    "            \n",
    "    def data(self, data):\n",
    "        global WFILE\n",
    "        if data is not None:\n",
    "            if self.is_accession and data != \"\":\n",
    "                WFILE.write(data)\n",
    "            \n",
    "    def close(self):\n",
    "        print('The xml file is closed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read xml file of variations from ClinVar\n",
    "# return dataframe and creates a subnode xml file\n",
    "def readClinVarVariationsXMLSpecific(input_path, accession):\n",
    "    print('Start parcing')\n",
    "    parser = etree.XMLParser(target=variationHandlerSpecific(accession))\n",
    "    etree.parse(input_path, parser)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # this column is to get a certain part of the xml file\n",
    "# WFILE = open('../data/subnode.xml', 'w')\n",
    "# xmlfile = '../data/ClinVarVariationRelease_00-latest_weekly.xml'\n",
    "# accession = 'VCV000424757'\n",
    "# readClinVarVariationsXMLSpecific(xmlfile, accession)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# makes dictionary of fasta sequences and np number \n",
    "# returns the dictionary\n",
    "def makeDictOfFasta(dictpath):\n",
    "    fasta_dict = {}\n",
    "    for root, d_names, file_names in os.walk(dictpath):\n",
    "        for filename in file_names:\n",
    "            fname = os.path.join(root, filename)\n",
    "            with open(fname, 'r') as f:\n",
    "                print('opened a fasta file')\n",
    "                np_num = ''\n",
    "                sequence = ''\n",
    "                for line in f:\n",
    "                    if line[0] == '>':\n",
    "                        if sequence != '':\n",
    "                            fasta_dict[np_num] = sequence\n",
    "                            np_num = ''\n",
    "                            sequence = ''                    \n",
    "                        i = 1\n",
    "                        while line[i] != ' ':\n",
    "                            np_num += line[i]\n",
    "                            i += 1\n",
    "                    else:\n",
    "                        line = line.strip('\\n')\n",
    "                        sequence += line\n",
    "    print(f'length of fasta dictionary: {len(fasta_dict)}')\n",
    "    return fasta_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "opened a fasta file\n",
      "opened a fasta file\n",
      "opened a fasta file\n",
      "opened a fasta file\n",
      "opened a fasta file\n",
      "opened a fasta file\n",
      "opened a fasta file\n",
      "length of fasta dictionary: 113645\n"
     ]
    }
   ],
   "source": [
    "fasta_dict = makeDictOfFasta('../fasta_sequences/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# crops fasta sequence\n",
    "# returns the cropped sequnece with a specified range\n",
    "def cropFASTA(sequence, location, reference, seqRange):\n",
    "    if location - 1 < len(sequence) and sequence[location - 1] == reference:\n",
    "        if location - 1 - seqRange <= 0:\n",
    "            proteinSeq = sequence[0:2 * seqRange + 1]\n",
    "        elif location - 1 + seqRange > len(sequence) - 1:\n",
    "            proteinSeq = sequence[0 if len(sequence) - 1 - 2 * seqRange <= 0 else len(sequence) - 1 - 2 * seqRange:len(sequence)]\n",
    "        else:\n",
    "            proteinSeq = sequence[location - 1 - seqRange : location + seqRange]\n",
    "        return proteinSeq\n",
    "    else:\n",
    "        return None "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def addFASTAfromDict(fasta_dict, df):\n",
    "    unfound = set()\n",
    "    seq_list = []\n",
    "    for index, row in df.iterrows():\n",
    "        mutation = row['missense_variation']\n",
    "        try:\n",
    "            ref = mutation[0]\n",
    "            try:\n",
    "                location = int(mutation[1:len(mutation)-1])\n",
    "            except:\n",
    "                location = int(mutation.split('_')[0][1:])\n",
    "            np_num = row['NP_accession']  # specify the column of np \n",
    "            sequence = fasta_dict[np_num]  # \n",
    "            seqRange = 12  # range of sequences to take\n",
    "            seq = cropFASTA(sequence, location, ref, seqRange) if sequence else None\n",
    "        except:\n",
    "            seq = None\n",
    "            unfound_np = row['NP_accession']\n",
    "            unfound.add(unfound_np)\n",
    "        seq_list.append(seq)\n",
    "    df['FASTA_window'] = seq_list\n",
    "    print(f'Unfound Sequences: {len(unfound)} {unfound}')\n",
    "    return df, unfound"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>EC_number</th>\n",
       "      <th>missense_variation</th>\n",
       "      <th>NC_accession</th>\n",
       "      <th>NP_accession</th>\n",
       "      <th>ClinVar_accession</th>\n",
       "      <th>Chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>referenceAllele</th>\n",
       "      <th>alternateAllele</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGA</td>\n",
       "      <td>aspartylglucosaminidase</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>3.5.1.26</td>\n",
       "      <td>G60D</td>\n",
       "      <td>NC_000004.12</td>\n",
       "      <td>NP_000018.2</td>\n",
       "      <td>VCV000000222</td>\n",
       "      <td>4</td>\n",
       "      <td>177440375.0</td>\n",
       "      <td>177440375.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DPYD</td>\n",
       "      <td>dihydropyrimidine dehydrogenase</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>1.3.1.2</td>\n",
       "      <td>R886H</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>NP_000101.2</td>\n",
       "      <td>VCV000000437</td>\n",
       "      <td>1</td>\n",
       "      <td>97098598.0</td>\n",
       "      <td>97098598.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PTS</td>\n",
       "      <td>6-pyruvoyltetrahydropterin synthase</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>4.2.3.12</td>\n",
       "      <td>R16C</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>NP_000308.1</td>\n",
       "      <td>VCV000000477</td>\n",
       "      <td>11</td>\n",
       "      <td>112226489.0</td>\n",
       "      <td>112226489.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PROC</td>\n",
       "      <td>protein C, inactivator of coagulation factors ...</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>3.4.21.69</td>\n",
       "      <td>P210L</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>NP_000303.1</td>\n",
       "      <td>VCV000000661</td>\n",
       "      <td>2</td>\n",
       "      <td>127426178.0</td>\n",
       "      <td>127426178.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARSB</td>\n",
       "      <td>arylsulfatase B</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>3.1.6.12</td>\n",
       "      <td>G137V</td>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>NP_000037.2</td>\n",
       "      <td>VCV000000877</td>\n",
       "      <td>5</td>\n",
       "      <td>78969095.0</td>\n",
       "      <td>78969095.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_ID                                          gene_name  \\\n",
       "0     AGA                            aspartylglucosaminidase   \n",
       "1    DPYD                    dihydropyrimidine dehydrogenase   \n",
       "2     PTS                6-pyruvoyltetrahydropterin synthase   \n",
       "3    PROC  protein C, inactivator of coagulation factors ...   \n",
       "4    ARSB                                    arylsulfatase B   \n",
       "\n",
       "    clinical_significance  EC_number missense_variation  NC_accession  \\\n",
       "0  uncertain significance   3.5.1.26               G60D  NC_000004.12   \n",
       "1  uncertain significance    1.3.1.2              R886H  NC_000001.11   \n",
       "2  uncertain significance   4.2.3.12               R16C  NC_000011.10   \n",
       "3  uncertain significance  3.4.21.69              P210L  NC_000002.12   \n",
       "4  uncertain significance   3.1.6.12              G137V  NC_000005.10   \n",
       "\n",
       "  NP_accession ClinVar_accession Chr        start         stop  \\\n",
       "0  NP_000018.2      VCV000000222   4  177440375.0  177440375.0   \n",
       "1  NP_000101.2      VCV000000437   1   97098598.0   97098598.0   \n",
       "2  NP_000308.1      VCV000000477  11  112226489.0  112226489.0   \n",
       "3  NP_000303.1      VCV000000661   2  127426178.0  127426178.0   \n",
       "4  NP_000037.2      VCV000000877   5   78969095.0   78969095.0   \n",
       "\n",
       "  referenceAllele alternateAllele strand  \n",
       "0               C               T      -  \n",
       "1               C               T      +  \n",
       "2               C               T      +  \n",
       "3               C               T      +  \n",
       "4               C               A      -  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_0 = pd.read_csv('../data/MM_enzyme.csv')\n",
    "df_0.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unfound Sequences: 23 {'NP_001138503.1', 'NP_000114.2', 'NP_001273003.1', 'NP_004251.3', 'NP_001180240.1', 'NP_000439.1', 'NP_001177317.1', 'NP_001202.4', 'NP_001017975.4', 'NP_001276896.1', 'NP_001041636.1', 'NP_065829.3', 'LRG_319p1', 'NP_000299.2', 'LRG_789p1', 'NP_945347.2', 'NP_001191354.1', 'NP_000342.2', 'NP_001017973.1', 'NP_004597.2', 'NP_000274.2', 'NP_000195.2', 'NP_000450.2'}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>EC_number</th>\n",
       "      <th>missense_variation</th>\n",
       "      <th>NC_accession</th>\n",
       "      <th>NP_accession</th>\n",
       "      <th>ClinVar_accession</th>\n",
       "      <th>Chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>referenceAllele</th>\n",
       "      <th>alternateAllele</th>\n",
       "      <th>strand</th>\n",
       "      <th>FASTA_window</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGA</td>\n",
       "      <td>aspartylglucosaminidase</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>3.5.1.26</td>\n",
       "      <td>G60D</td>\n",
       "      <td>NC_000004.12</td>\n",
       "      <td>NP_000018.2</td>\n",
       "      <td>VCV000000222</td>\n",
       "      <td>4</td>\n",
       "      <td>177440375.0</td>\n",
       "      <td>177440375.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>ASGGSALDAVESGCAMCEREQCDGS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DPYD</td>\n",
       "      <td>dihydropyrimidine dehydrogenase</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>1.3.1.2</td>\n",
       "      <td>R886H</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>NP_000101.2</td>\n",
       "      <td>VCV000000437</td>\n",
       "      <td>1</td>\n",
       "      <td>97098598.0</td>\n",
       "      <td>97098598.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>KKLPSFGPYLEQRKKIIAENKIRLK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PTS</td>\n",
       "      <td>6-pyruvoyltetrahydropterin synthase</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>4.2.3.12</td>\n",
       "      <td>R16C</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>NP_000308.1</td>\n",
       "      <td>VCV000000477</td>\n",
       "      <td>11</td>\n",
       "      <td>112226489.0</td>\n",
       "      <td>112226489.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>EGGGRRCQAQVSRRISFSASHRLYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PROC</td>\n",
       "      <td>protein C, inactivator of coagulation factors ...</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>3.4.21.69</td>\n",
       "      <td>P210L</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>NP_000303.1</td>\n",
       "      <td>VCV000000661</td>\n",
       "      <td>2</td>\n",
       "      <td>127426178.0</td>\n",
       "      <td>127426178.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>KRDTEDQEDQVDPRLIDGKMTRRGD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARSB</td>\n",
       "      <td>arylsulfatase B</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>3.1.6.12</td>\n",
       "      <td>G137V</td>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>NP_000037.2</td>\n",
       "      <td>VCV000000877</td>\n",
       "      <td>5</td>\n",
       "      <td>78969095.0</td>\n",
       "      <td>78969095.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>DEKLLPQLLKEAGYTTHMVGKWHLG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_ID                                          gene_name  \\\n",
       "0     AGA                            aspartylglucosaminidase   \n",
       "1    DPYD                    dihydropyrimidine dehydrogenase   \n",
       "2     PTS                6-pyruvoyltetrahydropterin synthase   \n",
       "3    PROC  protein C, inactivator of coagulation factors ...   \n",
       "4    ARSB                                    arylsulfatase B   \n",
       "\n",
       "    clinical_significance  EC_number missense_variation  NC_accession  \\\n",
       "0  uncertain significance   3.5.1.26               G60D  NC_000004.12   \n",
       "1  uncertain significance    1.3.1.2              R886H  NC_000001.11   \n",
       "2  uncertain significance   4.2.3.12               R16C  NC_000011.10   \n",
       "3  uncertain significance  3.4.21.69              P210L  NC_000002.12   \n",
       "4  uncertain significance   3.1.6.12              G137V  NC_000005.10   \n",
       "\n",
       "  NP_accession ClinVar_accession Chr        start         stop  \\\n",
       "0  NP_000018.2      VCV000000222   4  177440375.0  177440375.0   \n",
       "1  NP_000101.2      VCV000000437   1   97098598.0   97098598.0   \n",
       "2  NP_000308.1      VCV000000477  11  112226489.0  112226489.0   \n",
       "3  NP_000303.1      VCV000000661   2  127426178.0  127426178.0   \n",
       "4  NP_000037.2      VCV000000877   5   78969095.0   78969095.0   \n",
       "\n",
       "  referenceAllele alternateAllele strand               FASTA_window  \n",
       "0               C               T      -  ASGGSALDAVESGCAMCEREQCDGS  \n",
       "1               C               T      +  KKLPSFGPYLEQRKKIIAENKIRLK  \n",
       "2               C               T      +  EGGGRRCQAQVSRRISFSASHRLYS  \n",
       "3               C               T      +  KRDTEDQEDQVDPRLIDGKMTRRGD  \n",
       "4               C               A      -  DEKLLPQLLKEAGYTTHMVGKWHLG  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_1 = addFASTAfromDict(fasta_dict, df_0)\n",
    "df_1.to_csv('../data/MM_enzyme.csv', index = False, header = True)\n",
    "df_1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# some of the fasta sequences are not included in the fasta files so get them online\n",
    "def getFASTA(ls_np):\n",
    "    fasta_dict = {}\n",
    "    for np_num in ls_np:\n",
    "        handle = Entrez.efetch(db='protein', id=np_num, rettype='fasta', retmode='text', api_key='2959e9bc88ce27224b70cada27e5b58c6b09')\n",
    "        seq_record = SeqIO.read(handle, 'fasta')\n",
    "        sequence = seq_record.seq\n",
    "        fasta_dict[np_num] = sequence\n",
    "    return fasta_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "unfound_nps = ('NP_000114.2', 'NP_001017973.1', 'NP_000274.2','NP_000299.2', 'NP_001138503.1', 'NP_004251.3', 'NP_001276896.1', 'NP_001180240.1', 'NP_000195.2', 'NP_001017975.4', 'NP_001202.4', 'NP_945347.2', 'NP_001273003.1', 'NP_000439.1', 'NP_000450.2', 'NP_065829.3', 'NP_004597.2', 'NP_001191354.1', 'NP_001041636.1', 'NP_000342.2', 'NP_001177317.1')\n",
    "add_fasta = getFASTA(unfound_nps)\n",
    "fasta_dict.update(add_fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unfound Sequences: 2 {'LRG_319p1', 'LRG_789p1'}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>EC_number</th>\n",
       "      <th>missense_variation</th>\n",
       "      <th>NC_accession</th>\n",
       "      <th>NP_accession</th>\n",
       "      <th>ClinVar_accession</th>\n",
       "      <th>Chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>referenceAllele</th>\n",
       "      <th>alternateAllele</th>\n",
       "      <th>strand</th>\n",
       "      <th>FASTA_window</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGA</td>\n",
       "      <td>aspartylglucosaminidase</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>3.5.1.26</td>\n",
       "      <td>G60D</td>\n",
       "      <td>NC_000004.12</td>\n",
       "      <td>NP_000018.2</td>\n",
       "      <td>VCV000000222</td>\n",
       "      <td>4</td>\n",
       "      <td>177440375.0</td>\n",
       "      <td>177440375.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>ASGGSALDAVESGCAMCEREQCDGS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DPYD</td>\n",
       "      <td>dihydropyrimidine dehydrogenase</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>1.3.1.2</td>\n",
       "      <td>R886H</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>NP_000101.2</td>\n",
       "      <td>VCV000000437</td>\n",
       "      <td>1</td>\n",
       "      <td>97098598.0</td>\n",
       "      <td>97098598.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>KKLPSFGPYLEQRKKIIAENKIRLK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PTS</td>\n",
       "      <td>6-pyruvoyltetrahydropterin synthase</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>4.2.3.12</td>\n",
       "      <td>R16C</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>NP_000308.1</td>\n",
       "      <td>VCV000000477</td>\n",
       "      <td>11</td>\n",
       "      <td>112226489.0</td>\n",
       "      <td>112226489.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>EGGGRRCQAQVSRRISFSASHRLYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PROC</td>\n",
       "      <td>protein C, inactivator of coagulation factors ...</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>3.4.21.69</td>\n",
       "      <td>P210L</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>NP_000303.1</td>\n",
       "      <td>VCV000000661</td>\n",
       "      <td>2</td>\n",
       "      <td>127426178.0</td>\n",
       "      <td>127426178.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>KRDTEDQEDQVDPRLIDGKMTRRGD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARSB</td>\n",
       "      <td>arylsulfatase B</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>3.1.6.12</td>\n",
       "      <td>G137V</td>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>NP_000037.2</td>\n",
       "      <td>VCV000000877</td>\n",
       "      <td>5</td>\n",
       "      <td>78969095.0</td>\n",
       "      <td>78969095.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>DEKLLPQLLKEAGYTTHMVGKWHLG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_ID                                          gene_name  \\\n",
       "0     AGA                            aspartylglucosaminidase   \n",
       "1    DPYD                    dihydropyrimidine dehydrogenase   \n",
       "2     PTS                6-pyruvoyltetrahydropterin synthase   \n",
       "3    PROC  protein C, inactivator of coagulation factors ...   \n",
       "4    ARSB                                    arylsulfatase B   \n",
       "\n",
       "    clinical_significance  EC_number missense_variation  NC_accession  \\\n",
       "0  uncertain significance   3.5.1.26               G60D  NC_000004.12   \n",
       "1  uncertain significance    1.3.1.2              R886H  NC_000001.11   \n",
       "2  uncertain significance   4.2.3.12               R16C  NC_000011.10   \n",
       "3  uncertain significance  3.4.21.69              P210L  NC_000002.12   \n",
       "4  uncertain significance   3.1.6.12              G137V  NC_000005.10   \n",
       "\n",
       "  NP_accession ClinVar_accession Chr        start         stop  \\\n",
       "0  NP_000018.2      VCV000000222   4  177440375.0  177440375.0   \n",
       "1  NP_000101.2      VCV000000437   1   97098598.0   97098598.0   \n",
       "2  NP_000308.1      VCV000000477  11  112226489.0  112226489.0   \n",
       "3  NP_000303.1      VCV000000661   2  127426178.0  127426178.0   \n",
       "4  NP_000037.2      VCV000000877   5   78969095.0   78969095.0   \n",
       "\n",
       "  referenceAllele alternateAllele strand               FASTA_window  \n",
       "0               C               T      -  ASGGSALDAVESGCAMCEREQCDGS  \n",
       "1               C               T      +  KKLPSFGPYLEQRKKIIAENKIRLK  \n",
       "2               C               T      +  EGGGRRCQAQVSRRISFSASHRLYS  \n",
       "3               C               T      +  KRDTEDQEDQVDPRLIDGKMTRRGD  \n",
       "4               C               A      -  DEKLLPQLLKEAGYTTHMVGKWHLG  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_1 = addFASTAfromDict(fasta_dict, df_0)\n",
    "df_1.to_csv('../data/MM_enzyme.csv', index = False, header = True)\n",
    "df_1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NP_accession</th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>FASTA_window</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_000018.2</td>\n",
       "      <td>AGA</td>\n",
       "      <td>aspartylglucosaminidase</td>\n",
       "      <td>ASGGSALDAVESGCAMCEREQCDGS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NP_000101.2</td>\n",
       "      <td>DPYD</td>\n",
       "      <td>dihydropyrimidine dehydrogenase</td>\n",
       "      <td>KKLPSFGPYLEQRKKIIAENKIRLK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NP_000308.1</td>\n",
       "      <td>PTS</td>\n",
       "      <td>6-pyruvoyltetrahydropterin synthase</td>\n",
       "      <td>EGGGRRCQAQVSRRISFSASHRLYS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NP_000303.1</td>\n",
       "      <td>PROC</td>\n",
       "      <td>protein C, inactivator of coagulation factors ...</td>\n",
       "      <td>KRDTEDQEDQVDPRLIDGKMTRRGD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NP_000037.2</td>\n",
       "      <td>ARSB</td>\n",
       "      <td>arylsulfatase B</td>\n",
       "      <td>DEKLLPQLLKEAGYTTHMVGKWHLG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  NP_accession gene_ID                                          gene_name  \\\n",
       "0  NP_000018.2     AGA                            aspartylglucosaminidase   \n",
       "1  NP_000101.2    DPYD                    dihydropyrimidine dehydrogenase   \n",
       "2  NP_000308.1     PTS                6-pyruvoyltetrahydropterin synthase   \n",
       "3  NP_000303.1    PROC  protein C, inactivator of coagulation factors ...   \n",
       "4  NP_000037.2    ARSB                                    arylsulfatase B   \n",
       "\n",
       "                FASTA_window  \n",
       "0  ASGGSALDAVESGCAMCEREQCDGS  \n",
       "1  KKLPSFGPYLEQRKKIIAENKIRLK  \n",
       "2  EGGGRRCQAQVSRRISFSASHRLYS  \n",
       "3  KRDTEDQEDQVDPRLIDGKMTRRGD  \n",
       "4  DEKLLPQLLKEAGYTTHMVGKWHLG  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_1[['NP_accession', 'gene_ID', 'gene_name', 'FASTA_window']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make FASTA format text file from dataframe for blast search\n",
    "def makeFASTAfile(df, output_path):\n",
    "    subset = df[['NP_accession', 'gene_ID', 'gene_name', 'FASTA_window']]\n",
    "    tuples = [tuple(x) for x in subset.values]\n",
    "    with open(output_path, 'w') as f:\n",
    "        for tup in tuples:\n",
    "            line = '>' + '\\t'.join(tup[0:3]) + '\\n'\n",
    "            fasta = str(tup[3]) + '\\n'\n",
    "            f.write(line)\n",
    "            f.write(fasta) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "makeFASTAfile(df_1, '../data/fasta.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "def blastLocal(fasta_path, out_path, evalue=10.0, window_size=3):\n",
    "    start = datetime.datetime.now()  # for counting time necessary to run blast\n",
    "    cmd1 = '../ncbi/blast/bin/'\n",
    "    cmd2 = 'blastp' + ' '\\\n",
    "         + '-query ' + fasta_path + ' '\\\n",
    "         + '-db ' + cmd1 + 'pdbaa' + ' '\\\n",
    "         + '-evalue ' + str(evalue) + ' '\\\n",
    "         + '-outfmt ' + '5' + ' '\\\n",
    "         + '-out ' + out_path\n",
    "    cmd = cmd1 + cmd2\n",
    "    b_cmd = os.system(cmd)\n",
    "    end = datetime.datetime.now()\n",
    "    time = end - start\n",
    "    c = divmod(time.days * 86400 + time.seconds, 60)\n",
    "    print(cmd + ' : ran with exit code %d' %b_cmd)\n",
    "    print(f'Running Blast took {c[0]} minutes {c[1]} seconds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../ncbi/blast/bin/blastp -query ../data/fasta.txt -db ../ncbi/blast/bin/pdbaa -evalue 10.0 -outfmt 5 -out ../data/blast_result.xml : ran with exit code 0\n",
      "(21, 58)\n"
     ]
    }
   ],
   "source": [
    "fasta_path = '../data/fasta.txt'  # path for the fasta file\n",
    "out_path = '../data/blast_result.xml'  # path for the output file\n",
    "evalue = 10.0\n",
    "size = 3\n",
    "blastLocal(fasta_path, out_path, evalue, size)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "class BlastHandler(object):\n",
    "    def __init__(self):\n",
    "        self.dictlist = {'pdb_ID': [], 'BLAST_evalue': [], 'hit_from': [], 'hit_to': []}\n",
    "        self.is_hit_id = False\n",
    "        self.is_evalue = False\n",
    "        self.is_hit_from = False\n",
    "        self.is_hit_to = False\n",
    "        self.ct_iter = 0\n",
    "        self.ct = 0\n",
    "        \n",
    "    def start(self, tag, attrs):\n",
    "        if tag == 'Hit':  # there are a few <Hsp> tags within a <Hit> tag\n",
    "            self.ct_iter += 1\n",
    "        elif tag == 'Hit_id':\n",
    "            self.is_hit_id = True\n",
    "        elif tag == 'Hsp_evalue':\n",
    "            self.is_evalue = True\n",
    "        elif tag == 'Hsp_hit-from':\n",
    "            self.is_hit_from = True\n",
    "        elif tag == 'Hsp_hit-to':\n",
    "            self.is_hit_to = True\n",
    "            \n",
    "    def end(self, tag):\n",
    "        if tag == 'Iteration':\n",
    "            if self.ct_iter == 0:  # when there is no hit, add None to each list\n",
    "                self.dictlist['pdb_ID'].append(None)\n",
    "                self.dictlist['BLAST_evalue'].append(None)\n",
    "                self.dictlist['hit_from'].append(None)\n",
    "                self.dictlist['hit_to'].append(None)\n",
    "            self.ct_iter = 0\n",
    "            self.ct += 1\n",
    "            if self.ct % 10000 == 0:\n",
    "                print(f'coutner: {self.ct}')\n",
    "        elif tag == 'Hit_id':\n",
    "            self.is_hit_id = False\n",
    "        elif tag == 'Hsp_evalue':\n",
    "            self.is_evalue = False\n",
    "        elif tag == 'Hsp_hit-from':\n",
    "            self.is_hit_from = False\n",
    "        elif tag == 'Hsp_hit-to':\n",
    "            self.is_hit_to = False\n",
    "        elif tag == 'Hsp':\n",
    "            self.ct_iter += 1\n",
    "        \n",
    "    def data(self, data):\n",
    "        if self.ct_iter == 1:\n",
    "            if self.is_hit_id:\n",
    "                try:\n",
    "                    pdb = data.split('pdb|')[1]\n",
    "                except: \n",
    "                    print(f'Cannot split pdb {data}')\n",
    "                    pdb = None\n",
    "                self.dictlist['pdb_ID'].append(pdb)\n",
    "            elif self.is_evalue:\n",
    "                self.dictlist['BLAST_evalue'].append(data)\n",
    "            elif self.is_hit_from:\n",
    "                self.dictlist['hit_from'].append(data)\n",
    "            elif self.is_hit_to:\n",
    "                self.dictlist['hit_to'].append(data)\n",
    "    \n",
    "    def close(self):\n",
    "        print('Has completed parsing the blast_result xml')\n",
    "        print(f'Total Count: {self.ct}')\n",
    "        print(f'Length pdb: {len(self.dictlist[\"pdb_ID\"])}, evalue: {len(self.dictlist[\"BLAST_evalue\"])}, hit_from : {len(self.dictlist[\"hit_from\"])}, hit_to : {len(self.dictlist[\"hit_to\"])}')\n",
    "        return self.dictlist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readBlastXML(input_path):\n",
    "    print('Start parcing')\n",
    "    parser = etree.XMLParser(target=BlastHandler())\n",
    "    data = etree.parse(input_path, parser)\n",
    "    return data    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "def combineBlastResults(df, dictlist):\n",
    "    df['pdb_ID'] = dictlist['pdb_ID']\n",
    "    df['BLAST_evalue'] = dictlist['BLAST_evalue']\n",
    "    df['hit_from'] = dictlist['hit_from']\n",
    "    df['hit_to'] = dictlist['hit_to']\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start parcing\n",
      "coutner: 10000\n",
      "Length pdb: 10000, evalue: 10000, hit_from : 10000, hit_to : 10000\n",
      "coutner: 20000\n",
      "Length pdb: 20000, evalue: 20000, hit_from : 20000, hit_to : 20000\n",
      "coutner: 30000\n",
      "Length pdb: 30000, evalue: 30000, hit_from : 30000, hit_to : 30000\n",
      "coutner: 40000\n",
      "Length pdb: 40000, evalue: 40000, hit_from : 40000, hit_to : 40000\n",
      "coutner: 50000\n",
      "Length pdb: 50000, evalue: 50000, hit_from : 50000, hit_to : 50000\n",
      "coutner: 60000\n",
      "Length pdb: 60000, evalue: 60000, hit_from : 60000, hit_to : 60000\n",
      "coutner: 70000\n",
      "Length pdb: 70000, evalue: 70000, hit_from : 70000, hit_to : 70000\n",
      "Has completed parsing the blast_result xml\n",
      "Total Count: 70575\n",
      "Length pdb: 70575, evalue: 70575, hit_from : 70575, hit_to : 70575\n",
      "70575\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pdb_ID</th>\n",
       "      <th>BLAST_evalue</th>\n",
       "      <th>hit_from</th>\n",
       "      <th>hit_to</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1APY|A</td>\n",
       "      <td>6.44088e-11</td>\n",
       "      <td>25</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1GT8|A</td>\n",
       "      <td>1.87182e-09</td>\n",
       "      <td>874</td>\n",
       "      <td>898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3I2B|A</td>\n",
       "      <td>1.89206e-10</td>\n",
       "      <td>3</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4DT7|E</td>\n",
       "      <td>3.1355e-07</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1FSU|A</td>\n",
       "      <td>3.56958e-11</td>\n",
       "      <td>84</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pdb_ID BLAST_evalue hit_from hit_to\n",
       "0  1APY|A  6.44088e-11       25     49\n",
       "1  1GT8|A  1.87182e-09      874    898\n",
       "2  3I2B|A  1.89206e-10        3     24\n",
       "3  4DT7|E   3.1355e-07        1     19\n",
       "4  1FSU|A  3.56958e-11       84    108"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_dl = readBlastXML('../data/blast_result.xml')\n",
    "blast_df = pd.DataFrame(blast_dl)\n",
    "print(len(blast_dl['pdb_ID']))\n",
    "blast_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'blast_dl' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-37-70263ad2ee10>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mdf_1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'../data/MM_enzyme.csv'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mdf_2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcombineBlastResults\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf_1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mblast_dl\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mdf_2\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'../data/MM_enzyme_blast.csv'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mdf_2\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'blast_dl' is not defined"
     ]
    }
   ],
   "source": [
    "df_1 = pd.read_csv('../data/MM_enzyme.csv')\n",
    "df_2 = combineBlastResults(df_1, blast_dl)\n",
    "df_2.to_csv('../data/MM_enzyme_blast.csv', index = False, header = True)\n",
    "df_2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>EC_number</th>\n",
       "      <th>missense_variation</th>\n",
       "      <th>NP_accession</th>\n",
       "      <th>ClinVar_accession</th>\n",
       "      <th>Chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>referenceAllele</th>\n",
       "      <th>alternateAllele</th>\n",
       "      <th>FASTA_window</th>\n",
       "      <th>pdb_ID</th>\n",
       "      <th>BLAST_evalue</th>\n",
       "      <th>hit_from</th>\n",
       "      <th>hit_to</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGA</td>\n",
       "      <td>aspartylglucosaminidase</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>3.5.1.26</td>\n",
       "      <td>G60D</td>\n",
       "      <td>NP_000018.2</td>\n",
       "      <td>VCV000000222</td>\n",
       "      <td>4</td>\n",
       "      <td>177440375</td>\n",
       "      <td>177440375</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ASGGSALDAVESGCAMCEREQCDGS</td>\n",
       "      <td>1APY|A</td>\n",
       "      <td>6.440880e-11</td>\n",
       "      <td>25.0</td>\n",
       "      <td>49.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DPYD</td>\n",
       "      <td>dihydropyrimidine dehydrogenase</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>1.3.1.2</td>\n",
       "      <td>R886H</td>\n",
       "      <td>NP_000101.2</td>\n",
       "      <td>VCV000000437</td>\n",
       "      <td>1</td>\n",
       "      <td>97098598</td>\n",
       "      <td>97098598</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>KKLPSFGPYLEQRKKIIAENKIRLK</td>\n",
       "      <td>1GT8|A</td>\n",
       "      <td>1.871820e-09</td>\n",
       "      <td>874.0</td>\n",
       "      <td>898.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PTS</td>\n",
       "      <td>6-pyruvoyltetrahydropterin synthase</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>4.2.3.12</td>\n",
       "      <td>R16C</td>\n",
       "      <td>NP_000308.1</td>\n",
       "      <td>VCV000000477</td>\n",
       "      <td>11</td>\n",
       "      <td>112226489</td>\n",
       "      <td>112226489</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>EGGGRRCQAQVSRRISFSASHRLYS</td>\n",
       "      <td>3I2B|A</td>\n",
       "      <td>1.892060e-10</td>\n",
       "      <td>3.0</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PROC</td>\n",
       "      <td>protein C, inactivator of coagulation factors ...</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>3.4.21.69</td>\n",
       "      <td>P210L</td>\n",
       "      <td>NP_000303.1</td>\n",
       "      <td>VCV000000661</td>\n",
       "      <td>2</td>\n",
       "      <td>127426178</td>\n",
       "      <td>127426178</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>KRDTEDQEDQVDPRLIDGKMTRRGD</td>\n",
       "      <td>4DT7|E</td>\n",
       "      <td>3.135500e-07</td>\n",
       "      <td>1.0</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARSB</td>\n",
       "      <td>arylsulfatase B</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>3.1.6.12</td>\n",
       "      <td>G137V</td>\n",
       "      <td>NP_000037.2</td>\n",
       "      <td>VCV000000877</td>\n",
       "      <td>5</td>\n",
       "      <td>78969095</td>\n",
       "      <td>78969095</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>DEKLLPQLLKEAGYTTHMVGKWHLG</td>\n",
       "      <td>1FSU|A</td>\n",
       "      <td>3.569580e-11</td>\n",
       "      <td>84.0</td>\n",
       "      <td>108.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_ID                                          gene_name  \\\n",
       "0     AGA                            aspartylglucosaminidase   \n",
       "1    DPYD                    dihydropyrimidine dehydrogenase   \n",
       "2     PTS                6-pyruvoyltetrahydropterin synthase   \n",
       "3    PROC  protein C, inactivator of coagulation factors ...   \n",
       "4    ARSB                                    arylsulfatase B   \n",
       "\n",
       "    clinical_significance  EC_number missense_variation NP_accession  \\\n",
       "0  uncertain significance   3.5.1.26               G60D  NP_000018.2   \n",
       "1  uncertain significance    1.3.1.2              R886H  NP_000101.2   \n",
       "2  uncertain significance   4.2.3.12               R16C  NP_000308.1   \n",
       "3  uncertain significance  3.4.21.69              P210L  NP_000303.1   \n",
       "4  uncertain significance   3.1.6.12              G137V  NP_000037.2   \n",
       "\n",
       "  ClinVar_accession Chr      start       stop referenceAllele alternateAllele  \\\n",
       "0      VCV000000222   4  177440375  177440375               C               T   \n",
       "1      VCV000000437   1   97098598   97098598               C               T   \n",
       "2      VCV000000477  11  112226489  112226489               C               T   \n",
       "3      VCV000000661   2  127426178  127426178               C               T   \n",
       "4      VCV000000877   5   78969095   78969095               C               A   \n",
       "\n",
       "                FASTA_window  pdb_ID  BLAST_evalue  hit_from  hit_to  \n",
       "0  ASGGSALDAVESGCAMCEREQCDGS  1APY|A  6.440880e-11      25.0    49.0  \n",
       "1  KKLPSFGPYLEQRKKIIAENKIRLK  1GT8|A  1.871820e-09     874.0   898.0  \n",
       "2  EGGGRRCQAQVSRRISFSASHRLYS  3I2B|A  1.892060e-10       3.0    24.0  \n",
       "3  KRDTEDQEDQVDPRLIDGKMTRRGD  4DT7|E  3.135500e-07       1.0    19.0  \n",
       "4  DEKLLPQLLKEAGYTTHMVGKWHLG  1FSU|A  3.569580e-11      84.0   108.0  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# when you don't run but just import csv file\n",
    "df_2 = pd.read_csv('../data/MM_enzyme_blast.csv')\n",
    "df_2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "def orderColumns(df, col_names):\n",
    "    df_cols = set(df.columns)\n",
    "    if set(col_names).issubset(df_cols):\n",
    "            df_ordered = df.loc[:, col_names]\n",
    "            return df_ordered\n",
    "    else:\n",
    "        print('column names given include missing labels')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>EC_number</th>\n",
       "      <th>missense_variation</th>\n",
       "      <th>NP_accession</th>\n",
       "      <th>ClinVar_accession</th>\n",
       "      <th>Chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>referenceAllele</th>\n",
       "      <th>alternateAllele</th>\n",
       "      <th>FASTA_window</th>\n",
       "      <th>pdb_ID</th>\n",
       "      <th>BLAST_evalue</th>\n",
       "      <th>hit_from</th>\n",
       "      <th>hit_to</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGA</td>\n",
       "      <td>aspartylglucosaminidase</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>3.5.1.26</td>\n",
       "      <td>G60D</td>\n",
       "      <td>NP_000018.2</td>\n",
       "      <td>VCV000000222</td>\n",
       "      <td>4</td>\n",
       "      <td>177440375</td>\n",
       "      <td>177440375</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ASGGSALDAVESGCAMCEREQCDGS</td>\n",
       "      <td>1APY|A</td>\n",
       "      <td>6.44088e-11</td>\n",
       "      <td>25</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DPYD</td>\n",
       "      <td>dihydropyrimidine dehydrogenase</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>1.3.1.2</td>\n",
       "      <td>R886H</td>\n",
       "      <td>NP_000101.2</td>\n",
       "      <td>VCV000000437</td>\n",
       "      <td>1</td>\n",
       "      <td>97098598</td>\n",
       "      <td>97098598</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>KKLPSFGPYLEQRKKIIAENKIRLK</td>\n",
       "      <td>1GT8|A</td>\n",
       "      <td>1.87182e-09</td>\n",
       "      <td>874</td>\n",
       "      <td>898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PTS</td>\n",
       "      <td>6-pyruvoyltetrahydropterin synthase</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>4.2.3.12</td>\n",
       "      <td>R16C</td>\n",
       "      <td>NP_000308.1</td>\n",
       "      <td>VCV000000477</td>\n",
       "      <td>11</td>\n",
       "      <td>112226489</td>\n",
       "      <td>112226489</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>EGGGRRCQAQVSRRISFSASHRLYS</td>\n",
       "      <td>3I2B|A</td>\n",
       "      <td>1.89206e-10</td>\n",
       "      <td>3</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PROC</td>\n",
       "      <td>protein C, inactivator of coagulation factors ...</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>3.4.21.69</td>\n",
       "      <td>P210L</td>\n",
       "      <td>NP_000303.1</td>\n",
       "      <td>VCV000000661</td>\n",
       "      <td>2</td>\n",
       "      <td>127426178</td>\n",
       "      <td>127426178</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>KRDTEDQEDQVDPRLIDGKMTRRGD</td>\n",
       "      <td>4DT7|E</td>\n",
       "      <td>3.1355e-07</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARSB</td>\n",
       "      <td>arylsulfatase B</td>\n",
       "      <td>uncertain significance</td>\n",
       "      <td>3.1.6.12</td>\n",
       "      <td>G137V</td>\n",
       "      <td>NP_000037.2</td>\n",
       "      <td>VCV000000877</td>\n",
       "      <td>5</td>\n",
       "      <td>78969095</td>\n",
       "      <td>78969095</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>DEKLLPQLLKEAGYTTHMVGKWHLG</td>\n",
       "      <td>1FSU|A</td>\n",
       "      <td>3.56958e-11</td>\n",
       "      <td>84</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_ID                                          gene_name  \\\n",
       "0     AGA                            aspartylglucosaminidase   \n",
       "1    DPYD                    dihydropyrimidine dehydrogenase   \n",
       "2     PTS                6-pyruvoyltetrahydropterin synthase   \n",
       "3    PROC  protein C, inactivator of coagulation factors ...   \n",
       "4    ARSB                                    arylsulfatase B   \n",
       "\n",
       "    clinical_significance  EC_number missense_variation NP_accession  \\\n",
       "0  uncertain significance   3.5.1.26               G60D  NP_000018.2   \n",
       "1  uncertain significance    1.3.1.2              R886H  NP_000101.2   \n",
       "2  uncertain significance   4.2.3.12               R16C  NP_000308.1   \n",
       "3  uncertain significance  3.4.21.69              P210L  NP_000303.1   \n",
       "4  uncertain significance   3.1.6.12              G137V  NP_000037.2   \n",
       "\n",
       "  ClinVar_accession Chr      start       stop referenceAllele alternateAllele  \\\n",
       "0      VCV000000222   4  177440375  177440375               C               T   \n",
       "1      VCV000000437   1   97098598   97098598               C               T   \n",
       "2      VCV000000477  11  112226489  112226489               C               T   \n",
       "3      VCV000000661   2  127426178  127426178               C               T   \n",
       "4      VCV000000877   5   78969095   78969095               C               A   \n",
       "\n",
       "                FASTA_window  pdb_ID BLAST_evalue hit_from hit_to  \n",
       "0  ASGGSALDAVESGCAMCEREQCDGS  1APY|A  6.44088e-11       25     49  \n",
       "1  KKLPSFGPYLEQRKKIIAENKIRLK  1GT8|A  1.87182e-09      874    898  \n",
       "2  EGGGRRCQAQVSRRISFSASHRLYS  3I2B|A  1.89206e-10        3     24  \n",
       "3  KRDTEDQEDQVDPRLIDGKMTRRGD  4DT7|E   3.1355e-07        1     19  \n",
       "4  DEKLLPQLLKEAGYTTHMVGKWHLG  1FSU|A  3.56958e-11       84    108  "
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns = ('gene_ID', 'gene_name', 'clinical_significance', 'EC_number', 'missense_variation', 'NP_accession', 'ClinVar_accession', 'gnomAD_AF', 'SIFT_prediction', 'CADD_score', 'Chr', 'start', 'stop', 'referenceAllele', 'alternateAllele', 'FASTA_window', 'pdb_ID', 'BLAST_evalue', 'hit_from', 'hit_to')\n",
    "columns2 = ('gene_ID', 'gene_name', 'clinical_significance', 'EC_number', 'missense_variation', 'NP_accession', 'ClinVar_accession', 'Chr', 'start', 'stop', 'referenceAllele', 'alternateAllele', 'FASTA_window', 'pdb_ID', 'BLAST_evalue', 'hit_from', 'hit_to')\n",
    "df_3 = orderColumns(df_2, columns2)\n",
    "df_3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def makeInputForVep(df, output_path):\n",
    "    df_sub = df[['Chr', 'start', 'stop', 'referenceAllele', 'alternateAllele']]\n",
    "    df_sub.to_csv(output_path, sep='\\t', index=False, header=False)\n",
    "    return df_sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>referenceAllele</th>\n",
       "      <th>alternateAllele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>177440375</td>\n",
       "      <td>177440375</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>97098598</td>\n",
       "      <td>97098598</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11</td>\n",
       "      <td>112226489</td>\n",
       "      <td>112226489</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>127426178</td>\n",
       "      <td>127426178</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>78969095</td>\n",
       "      <td>78969095</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chr      start       stop referenceAllele alternateAllele\n",
       "0   4  177440375  177440375               C               T\n",
       "1   1   97098598   97098598               C               T\n",
       "2  11  112226489  112226489               C               T\n",
       "3   2  127426178  127426178               C               T\n",
       "4   5   78969095   78969095               C               A"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vep_input_path = '../data/vep_input.txt'\n",
    "df_2_sub = makeInputForVep(df_2, vep_input_path)\n",
    "df_2_sub.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "def vepLocal(input_path, output_path):\n",
    "    start = datetime.datetime.now()  # for counting time necessary to run blast\n",
    "    cmd1 = '../../ensembl-vep/'  # specify directory\n",
    "    cmd2 = './vep' + ' '\\\n",
    "         + '-i ' + input_path + ' '\\\n",
    "         + '-o ' + output_path + ' '\\\n",
    "         + '--cache ' + '--af_gnomad '\\\n",
    "         + '--appris ' + '--canonical '\\\n",
    "         + '--sift p ' + '--polyphen p '\\\n",
    "         + '--no_check_variants_order '\\\n",
    "         + '--flag_pick ' + '--tab '\\\n",
    "         + '--force_overwrite'\n",
    "    cmd = cmd1 + cmd2\n",
    "    vep_cmd = os.system(cmd)\n",
    "    end = datetime.datetime.now()\n",
    "    time = end - start\n",
    "    c = divmod(time.days * 86400 + time.seconds, 60)\n",
    "    print(cmd + ' : ran with exit code %d' %vep_cmd)\n",
    "    print(f'Running Vep took {c[0]} minutes {c[1]} seconds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "../../ensembl-vep/./vep -i ../data/vep_input.txt -o ../data/vep_output.txt --cache --af_gnomad --appris --canonical --sift p --polyphen p --flag_pick --tab --force_overwrite : ran with exit code 2\n",
      "Running Vep took 0 minutes 8 seconds\n"
     ]
    }
   ],
   "source": [
    "vep_output_path = '../data/vep_output.txt'\n",
    "vepLocal(vep_input_path, vep_output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../../ensembl-vep/./vep -i ../data/inputForVep_short.txt -o ../data/vep_output_short.txt --cache --af_gnomad --appris --canonical --sift p --polyphen p --no_check_variants_order --flag_pick --tab --force_overwrite : ran with exit code 0\n",
      "Running Vep took 0 minutes 41 seconds\n"
     ]
    }
   ],
   "source": [
    "# vep demo\n",
    "vep_output_path_demo = '../data/vep_output_short.txt'\n",
    "vepLocal('../data/inputForVep_short.txt', vep_output_path_demo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cropVepOutput(input_path, output_path):\n",
    "    cmd = \"cat \" + input_path + \" | \"\\\n",
    "        + \"grep -v '##'\" + \" | \"\\\n",
    "        + \"sed 's/^#\\(.*\\)/\\\\1/'\" +  \" >| \"\\\n",
    "        + output_path\n",
    "    crop_cmd = os.system(cmd)\n",
    "    print(cmd + ' : ran with exit code %d' %crop_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "crop_vep_path = '../data/vep_output_cropped.txt'\n",
    "cropVepOutput(vep_output_path, crop_vep_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cat ../data/vep_output_short.txt | grep -v '##' | sed 's/^#\\(.*\\)/\\1/' >| ../data/vep_output_short_cropped.txt : ran with exit code 0\n"
     ]
    }
   ],
   "source": [
    "# vep demo\n",
    "crop_vep_path_demo = '../data/vep_output_short_cropped.txt'\n",
    "cropVepOutput(vep_output_path_demo, crop_vep_path_demo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>...</th>\n",
       "      <th>gnomAD_AMR_AF</th>\n",
       "      <th>gnomAD_ASJ_AF</th>\n",
       "      <th>gnomAD_EAS_AF</th>\n",
       "      <th>gnomAD_FIN_AF</th>\n",
       "      <th>gnomAD_NFE_AF</th>\n",
       "      <th>gnomAD_OTH_AF</th>\n",
       "      <th>gnomAD_SAS_AF</th>\n",
       "      <th>CLIN_SIG</th>\n",
       "      <th>SOMATIC</th>\n",
       "      <th>PHENO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>155238261</td>\n",
       "      <td>1:155238261</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000177628</td>\n",
       "      <td>ENST00000327247</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>867</td>\n",
       "      <td>634</td>\n",
       "      <td>212</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>155238261</td>\n",
       "      <td>1:155238261</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000177628</td>\n",
       "      <td>ENST00000368373</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>771</td>\n",
       "      <td>634</td>\n",
       "      <td>212</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>155238261</td>\n",
       "      <td>1:155238261</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000177628</td>\n",
       "      <td>ENST00000427500</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>650</td>\n",
       "      <td>487</td>\n",
       "      <td>163</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>155238261</td>\n",
       "      <td>1:155238261</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000177628</td>\n",
       "      <td>ENST00000428024</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>713</td>\n",
       "      <td>373</td>\n",
       "      <td>125</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>155238261</td>\n",
       "      <td>1:155238261</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000236675</td>\n",
       "      <td>ENST00000440904</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Uploaded_variation     Location Allele             Gene          Feature  \\\n",
       "0           155238261  1:155238261      T  ENSG00000177628  ENST00000327247   \n",
       "1           155238261  1:155238261      T  ENSG00000177628  ENST00000368373   \n",
       "2           155238261  1:155238261      T  ENSG00000177628  ENST00000427500   \n",
       "3           155238261  1:155238261      T  ENSG00000177628  ENST00000428024   \n",
       "4           155238261  1:155238261      T  ENSG00000236675  ENST00000440904   \n",
       "\n",
       "  Feature_type              Consequence cDNA_position CDS_position  \\\n",
       "0   Transcript         missense_variant           867          634   \n",
       "1   Transcript         missense_variant           771          634   \n",
       "2   Transcript         missense_variant           650          487   \n",
       "3   Transcript         missense_variant           713          373   \n",
       "4   Transcript  downstream_gene_variant             -            -   \n",
       "\n",
       "  Protein_position  ... gnomAD_AMR_AF gnomAD_ASJ_AF gnomAD_EAS_AF  \\\n",
       "0              212  ...             -             -             -   \n",
       "1              212  ...             -             -             -   \n",
       "2              163  ...             -             -             -   \n",
       "3              125  ...             -             -             -   \n",
       "4                -  ...             -             -             -   \n",
       "\n",
       "  gnomAD_FIN_AF gnomAD_NFE_AF  gnomAD_OTH_AF gnomAD_SAS_AF  \\\n",
       "0             -             -              -             -   \n",
       "1             -             -              -             -   \n",
       "2             -             -              -             -   \n",
       "3             -             -              -             -   \n",
       "4             -             -              -             -   \n",
       "\n",
       "                 CLIN_SIG SOMATIC PHENO  \n",
       "0  uncertain_significance       -     1  \n",
       "1  uncertain_significance       -     1  \n",
       "2  uncertain_significance       -     1  \n",
       "3  uncertain_significance       -     1  \n",
       "4  uncertain_significance       -     1  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_vep = pd.read_csv(crop_vep_path_demo, sep='\\t', header=0)\n",
    "df_vep.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>...</th>\n",
       "      <th>gnomAD_AMR_AF</th>\n",
       "      <th>gnomAD_ASJ_AF</th>\n",
       "      <th>gnomAD_EAS_AF</th>\n",
       "      <th>gnomAD_FIN_AF</th>\n",
       "      <th>gnomAD_NFE_AF</th>\n",
       "      <th>gnomAD_OTH_AF</th>\n",
       "      <th>gnomAD_SAS_AF</th>\n",
       "      <th>CLIN_SIG</th>\n",
       "      <th>SOMATIC</th>\n",
       "      <th>PHENO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>155238261</td>\n",
       "      <td>1:155238261</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000177628</td>\n",
       "      <td>ENST00000327247</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>867</td>\n",
       "      <td>634</td>\n",
       "      <td>212</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>100206558</td>\n",
       "      <td>1:100206558</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000137992</td>\n",
       "      <td>ENST00000370132</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1110</td>\n",
       "      <td>1096</td>\n",
       "      <td>366</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>uncertain_significance</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>100196274</td>\n",
       "      <td>1:100196274</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000137992</td>\n",
       "      <td>ENST00000370132</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1444</td>\n",
       "      <td>1430</td>\n",
       "      <td>477</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8.809e-06</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>uncertain_significance,pathogenic</td>\n",
       "      <td>-</td>\n",
       "      <td>1,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>226888945</td>\n",
       "      <td>1:226888945</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000143801</td>\n",
       "      <td>ENST00000366783</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1066</td>\n",
       "      <td>683</td>\n",
       "      <td>228</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.759e-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>not_provided</td>\n",
       "      <td>-</td>\n",
       "      <td>1,1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>94080562</td>\n",
       "      <td>1:94080562</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000198691</td>\n",
       "      <td>ENST00000370225</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1118</td>\n",
       "      <td>1015</td>\n",
       "      <td>339</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.763e-05</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>conflicting_interpretations_of_pathogenicity</td>\n",
       "      <td>-</td>\n",
       "      <td>1,1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Uploaded_variation     Location Allele             Gene          Feature  \\\n",
       "0            155238261  1:155238261      T  ENSG00000177628  ENST00000327247   \n",
       "16           100206558  1:100206558      G  ENSG00000137992  ENST00000370132   \n",
       "17           100196274  1:100196274      C  ENSG00000137992  ENST00000370132   \n",
       "20           226888945  1:226888945      T  ENSG00000143801  ENST00000366783   \n",
       "28            94080562   1:94080562      C  ENSG00000198691  ENST00000370225   \n",
       "\n",
       "   Feature_type       Consequence cDNA_position CDS_position Protein_position  \\\n",
       "0    Transcript  missense_variant           867          634              212   \n",
       "16   Transcript  missense_variant          1110         1096              366   \n",
       "17   Transcript  missense_variant          1444         1430              477   \n",
       "20   Transcript  missense_variant          1066          683              228   \n",
       "28   Transcript  missense_variant          1118         1015              339   \n",
       "\n",
       "    ... gnomAD_AMR_AF gnomAD_ASJ_AF gnomAD_EAS_AF gnomAD_FIN_AF gnomAD_NFE_AF  \\\n",
       "0   ...             -             -             -             -             -   \n",
       "16  ...             -             -             -             -             -   \n",
       "17  ...             0             0             0             0     8.809e-06   \n",
       "20  ...             0             0             0             0     1.759e-05   \n",
       "28  ...             0             0             0             0     1.763e-05   \n",
       "\n",
       "    gnomAD_OTH_AF gnomAD_SAS_AF                                      CLIN_SIG  \\\n",
       "0               -             -                        uncertain_significance   \n",
       "16              -             -                        uncertain_significance   \n",
       "17              0             0             uncertain_significance,pathogenic   \n",
       "20              0             0                                  not_provided   \n",
       "28              0             0  conflicting_interpretations_of_pathogenicity   \n",
       "\n",
       "   SOMATIC PHENO  \n",
       "0        -     1  \n",
       "16       -     1  \n",
       "17       -   1,1  \n",
       "20       -   1,1  \n",
       "28       -   1,1  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_vep_pick = df_vep.loc[df_vep['PICK'] == '1']\n",
    "df_vep_pick.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ./vep -i inputForVep_short.txt -o outputForVep_short.txt --cache --af_gnomad --appris --canonical --sift p --polyphen p --force_overwrite --tab --flag_pick --no_check_variants_order\n",
    "# ./vep -i inputForVep.txt -o outputForVep.txt --cache --af_gnomad --appris --canonical --sift p --polyphen p --force_overwrite --tab --flag_pick"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://cadd.gs.washington.edu/api/v1.0/GRCh38-v1.4/22:44044001_T_A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #Specify score file link\n",
    "# SCORE_FILE=http://krishna.gs.washington.edu/download/CADD/v1.4/GRCh37/whole_genome_SNVs_inclAnno.tsv.gz\n",
    "# #Download Tabix Index\n",
    "# INDEX=IndexFile\n",
    "# wget -c $SCORE_FILE.tbi -O $INDEX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[NbConvertApp] Converting notebook VersUS.ipynb to script\n",
      "[NbConvertApp] Writing 21911 bytes to VersUS.py\n"
     ]
    }
   ],
   "source": [
    "!jupyter nbconvert --to script VersUS.ipynb\n",
    "with open('VersUS.py', 'r') as f:\n",
    "    lines = f.readlines()\n",
    "with open('VersUS.py', 'w') as f:\n",
    "    for line in lines:\n",
    "        if 'nbconvert --to script' in line:\n",
    "            break\n",
    "        else:\n",
    "            f.write(line)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
